# Filtering arguments for annotating edits
# Aaron Wagen
# Nov 2022


# Order of factors for each category from most preferenced to least

filtering_args <- list(alu_repeat_levels = c("Simple_repeat", "SINE/Alu", "SINE/MIR", "LINE/L1", "LINE/L2", "LTRs", "DNAs", "Retroposons/SVA", "RNAs", "Other SINE", "Other LINE", "Other"),
                       gtf_genic_type = c("exon", "intron", "intergenic"),
                       gtf_genic_biotype = c("protein_coding",
                                             "processed_transcript",
                                             "lincRNA",
                                             "antisense",
                                             "macro_lncRNA",
                                             "sense_intronic",
                                             "sense_overlapping",
                                             "3prime_overlapping_ncRNA",
                                             "transcribed_processed_pseudogene",
                                             "transcribed_unitary_pseudogene",
                                             "bidirectional_promoter_lncRNA",
                                             "IG_C_gene",
                                             "IG_C_pseudogene",
                                             "IG_D_gene",
                                             "IG_J_gene",
                                             "IG_J_pseudogene",
                                             "IG_pseudogene",
                                             "IG_V_gene",
                                             "IG_V_pseudogene",
                                             "miRNA",
                                             "misc_RNA",
                                             "Mt_rRNA",
                                             "Mt_tRNA",
                                             "non_coding",
                                             "polymorphic_pseudogene",
                                             "pseudogene",
                                             "processed_pseudogene",
                                             "ribozyme",
                                             "rRNA",
                                             "scaRNA",
                                             "scRNA",
                                             "snoRNA",
                                             "snRNA",
                                             "sRNA" ,
                                             "TEC",
                                             "TR_C_gene",
                                             "TR_D_gene",
                                             "TR_J_gene",
                                             "TR_J_pseudogene",
                                             "TR_V_gene",
                                             "TR_V_pseudogene",
                                             "transcribed_unprocessed_pseudogene",
                                             "translated_processed_pseudogene",
                                             "unitary_pseudogene",
                                             "unprocessed_pseudogene",
                                             "vaultRNA"),
                       vep_impact = c("HIGH", "MODERATE", "LOW", "MODIFIER"),
                       vep_consequence = c("transcript_ablation", # As per http://www.ensembl.org/info/genome/variation/prediction/predicted_data.html
                                           "splice_acceptor_variant",
                                           "splice_donor_variant",
                                           "stop_gained",
                                           "frameshift_variant",
                                           "stop_lost",
                                           "start_lost",
                                           "transcript_amplification",
                                           "inframe_insertion",
                                           "inframe_deletion",
                                           "missense_variant",
                                           "protein_altering_variant",
                                           "splice_region_variant",
                                           "incomplete_terminal_codon_variant",
                                           "start_retained_variant",
                                           "stop_retained_variant",
                                           "synonymous_variant",
                                           "coding_sequence_variant", # These are initiator variants or terminator veriants (protein altering variant and sysnonymous variants are also coding sequence variants)
                                           "mature_miRNA_variant",
                                           "5_prime_UTR_variant",
                                           "3_prime_UTR_variant",
                                           "non_coding_transcript_exon_variant",
                                           "intron_variant",
                                           "NMD_transcript_variant",
                                           "non_coding_transcript_variant",
                                           "upstream_gene_variant",
                                           "downstream_gene_variant",
                                           "TFBS_ablation",
                                           "TFBS_amplification",
                                           "TF_binding_site_variant",
                                           "regulatory_region_ablation",
                                           "regulatory_region_amplification",
                                           "feature_elongation",
                                           "regulatory_region_variant",
                                           "feature_truncation",
                                           "intergenic_variant"),
                       vep_biotype = c("protein_coding",
                                       "lincRNA",
                                       "macro_lncRNA",
                                       "bidirectional_promoter_lncRNA",
                                       "antisense",
                                       "sense_intronic",
                                       "sense_overlapping",
                                       "non_stop_decay",
                                       "nonsense_mediated_decay",
                                       "3prime_overlapping_ncRNA",
                                       "retained_intron",
                                       "processed_transcript",
                                       "translated_processed_pseudogene",
                                       "transcribed_processed_pseudogene",
                                       "transcribed_unitary_pseudogene",
                                       "transcribed_unprocessed_pseudogene",
                                       "processed_pseudogene",
                                       "IG_V_gene",
                                       "IG_C_gene",
                                       "IG_C_pseudogene",
                                       "IG_J_gene",
                                       "IG_V_pseudogene",
                                       "miRNA",
                                       "misc_RNA",
                                       "Mt_rRNA",
                                       "Mt_tRNA",
                                       "non_coding",
                                       "polymorphic_pseudogene",
                                       "pseudogene",
                                       "ribozyme",
                                       "rRNA",
                                       "scaRNA",
                                       "scRNA",
                                       "snoRNA",
                                       "snRNA",
                                       "sRNA",
                                       "TEC",
                                       "TR_C_gene",
                                       "TR_V_gene",
                                       "TR_D_gene",
                                       "TR_J_gene",
                                       "TR_V_pseudogene",
                                       "unitary_pseudogene",
                                       "unprocessed_pseudogene"),
                       vep_canonical = c("YES", "-"),
                       vep_TSL = c(1:5, "-"),
                       sample_levels = c("astro_nt", "astro_asyn", "cocult_nt", "cocult_asyn", "neuron_nt", "neuron_asyn"),
                       sample_labels = c("Astrocyte no treatment", "Astrocyte a-syn", "Coculture no treatment", "Coculture a-syn", "Neuron no treatment", "Neuron a-syn"),
                       vep_consequence_plots <-  list("3 prime UTR variant" = "3_prime_UTR_variant",
                                                      "Downstream variant" =  "downstream_gene_variant",
                                                      "Intron variant" = "intron_variant",
                                                      "Missense variant"  = c("inframe_insertion",
                                                                              "inframe_deletion",
                                                                              "missense_variant",
                                                                              "protein_altering_variant"),
                                                      "Ablation / Frameshift" = c("transcript_ablation",
                                                                                  "frameshift_variant"),
                                                      "Splice variant" = c("splice_acceptor_variant",
                                                                           "splice_donor_variant",
                                                                           "splice_region_variant",
                                                                           "splice_donor_region_variant"),
                                                      "Start/stop variant" = c("stop_lost",
                                                                            "start_lost",
                                                                            "stop_gained",
                                                                            "start_retained_variant",
                                                                            "stop_retained_variant"),
                                                      "Synonymous variant" =  "synonymous_variant",
                                                      "Coding sequence variant" = c("coding_sequence_variant",
                                                                                    "transcript_amplification"),
                                                      "5 prime UTR variant" = "5_prime_UTR_variant",
                                                      "Upstream variant" =  "upstream_gene_variant",
                                                      "Other" = c("incomplete_terminal_codon_variant",
                                                                  "mature_miRNA_variant",
                                                                  "non_coding_transcript_exon_variant",
                                                                  "non_coding_transcript_variant",
                                                                  "incomplete_terminal_codon_variant",
                                                                  "feature_elongation",
                                                                  "regulatory_region_variant",
                                                                  "feature_truncation"), 
                                                      "miRNA variant" = "mature_miRNA_variant",
                                                      "Transcription factor / Regulation " = c("TFBS_ablation",
                                                                                               "TFBS_amplification",
                                                                                               "TF_binding_site_variant",
                                                                                               "regulatory_region_ablation",
                                                                                               "regulatory_region_amplification"),
                                                      "Intergenic" = "intergenic_variant"
                       ),
                       vep_levels_BIOTYPE <- list("Protein coding" = "protein_coding",
                                                  "LINC" = c("lincRNA",
                                                             "macro_lncRNA",
                                                             "bidirectional_promoter_lncRNA"),
                                                  "Antisense" = "antisense",
                                                  "Sense Intronic/overlapping" = c( "sense_intronic", "sense_overlapping"),
                                                  "Non stop decay" = "non_stop_decay",
                                                  "Non coding" =  "non_coding",
                                                  "Nonsense mediated decay" = "nonsense_mediated_decay",
                                                  "3prime_overlapping_ncRNA" = "3prime_overlapping_ncRNA" ,
                                                  "Retained_intron" = "retained_intron",
                                                  "processed_transcript" ="processed_transcript" ,
                                                  "Pseudogene" = c("translated_processed_pseudogene",
                                                                   "transcribed_processed_pseudogene",
                                                                   "transcribed_unitary_pseudogene",
                                                                   "transcribed_unprocessed_pseudogene",
                                                                   "processed_pseudogene",
                                                                   "polymorphic_pseudogene",
                                                                   "pseudogene" ,
                                                                   "unitary_pseudogene",
                                                                   "unprocessed_pseudogene"),
                                                  "Immunoglobulin" = c("IG_V_gene",
                                                                       "IG_C_gene",
                                                                       "IG_C_pseudogene",
                                                                       "IG_J_gene",
                                                                       "IG_V_pseudogene"),
                                                  "T Cell Receptor" = c("TR_C_gene",
                                                                        "TR_V_gene",
                                                                        "TR_D_gene",
                                                                        "TR_J_gene",
                                                                        "TR_V_pseudogene"),
                                                  "miRNA" = "miRNA",
                                                  "Splicing variant" = c("splice_region_variant",
                                                                         "splice_donor_variant",
                                                                         "splice_acceptor_variant"),
                                                  "Mitochondrial" = c( "Mt_rRNA", "Mt_tRNA" ),
                                                  "Other" = c("scaRNA",
                                                              "scRNA",
                                                              "snoRNA",
                                                              "snRNA",
                                                              "sRNA",
                                                              "misc_RNA",
                                                              "ribozyme",
                                                              "rRNA",
                                                              "TEC",
                                                              "non_coding")
                       ), 
                       genomic_repeat_levels <- list("Simple_repeat" = "Simple_repeat",
                                             "LINE-L1" = "L1",
                                             "LINE-L1" = "L2",
                                             "LINE-Other" = c("CR1",
                                                              "RTE-X",
                                                              "RTE-BovB",
                                                              "Penelope",
                                                              "Dong-R4"),
                                             "DNA" = c("hAT-Charlie",
                                                        "hAT",
                                                        "hAT-Tip100",
                                                        "TcMar-Tigger",
                                                        "MULE-MuDR",
                                                        "hAT-Blackjack",
                                                        "hAT-Ac",
                                                        "TcMar-Tc2",
                                                        "PiggyBac",
                                                        "hAT?",
                                                        "TcMar-Mariner",
                                                        "hAT-Tip100?",
                                                        "DNA",
                                                        "PiggyBac?",
                                                        "TcMar?",
                                                        "TcMar",
                                                        "hAT-Tag1",
                                                        "Merlin",
                                                        "TcMar-Pogo",
                                                        "PIF-Harbinger",
                                                       "DNA?"),
                                             "SINE-Alu" = "Alu",
                                             "SINE-MIR" = "MIR",
                                             "SINE-Other" = c("tRNA",
                                                              "tRNA-Deu",
                                                              "5S-Deu-L2",
                                                              "tRNA-RTE",
                                                              "SINE?"),
                                             "ERV" = c("ERVL-MaLR",
                                                       "ERVL",
                                                       "ERV1",
                                                       "ERVL?",
                                                       "ERV1?",
                                                       "ERVK"),
                                             "LTR" = c("LTR",
                                                       "Gypsy",
                                                       "Gypsy?",
                                                       "LTR?"),
                                             "Other" = c("telo",
                                                         "Satellite",
                                                         "centr",
                                                         "acro",
                                                         "Low_complexity",
                                                         "snRNA",
                                                         "SVA",
                                                         "srpRNA",
                                                         "rRNA",
                                                         "Unknown",
                                                         "Helitron",
                                                         "scRNA",
                                                         "RNA",
                                                         "Helitron?",
                                                         "tRNA")
                       )
)




# Old consequence filtering args
# "missense",
# "coding_sequence",  # Terminator and initiator changes
# "splicing",
# "synonymous",
# "3_prime",
# "5_prime",
# "non_coding_exon",
# "intronic",
# "down/upstream",
# "intergenic",
# "miRNA"