---
title: "Processing astrocyte-neuron samples"
author: 
- name: "Aaron Wagen"
  affiliation: UCL
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  bookdown::html_document2:
    figure_caption: yes
    code_folding: show
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    number_sections: true
    css: styles.css
  md_document:
    variant: markdown_github
    toc: true
    number_sections: true
always_allow_html: true
link-citations: true
notes-after-punctuation: false
---


```{r setup, include = FALSE}

library(here) # For project-specific paths
library(tidyverse)
library(GenomicRanges)
library(plyranges)
library(patchwork)
library(paletteer)
library(ggtranscript)
library(pander)
library(knitr)
library(kableExtra)
library(data.table)
library(EWCE)
library(broom)
library(tidycat)
library(data.table)



knitr::opts_chunk$set(echo = T, warning = F, message = F)
options(bitmapType='cairo')


here::i_am("docs/01_astrocyte_analysis.Rmd")

# Load functions and filters
source(here::here("functions_args",
                  "r_functions.R"))



options(bitmapType = 'cairo',
        scipen = 7) # Prints any number less than 7 digits in normal (non scientific) notation

# Set defaults for plots for markdown
theme_aw <- theme_bw(base_family = "Helvetica",
           base_size = 12) + 
  theme(panel.grid.major.x = element_blank(),
        strip.background = element_blank(),
        legend.position = "right",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.y = element_text(vjust = 0.6),
        panel.spacing = unit(0.1, "lines"))

# Set defaults for plots for publication
STRIP_TEXT <- 12
SIZE_AXIS_TITLE <- 12
SIZE_AXIS_TEXT <-12 
SIZE_LEGEND_TITLE <-16 
SIZE_LEGEND_TEXT <- 12
SIZE_PLOT_TITLE <- 16
SIZE_PLOT_SUBTITLE <- 12
#set the theme with base font
theme_plot <- function( base_family = "Helvetica") {
  theme_bw( base_family = base_family ) +
    theme(legend.position = "right") +
    theme(panel.grid.major = element_blank(), 
          #      panel.background = element_blank(), 
          #panel.border = element_blank(),
          #       axis.line = element_line(colour = "black"), #clear the background
          strip.background = element_blank(),
          strip.text = element_text(colour = "black",
                                    size = STRIP_TEXT),
          axis.title  = element_text(color="black", 
                                     size=SIZE_AXIS_TITLE),
          axis.text  = element_text( color="black", 
                                     size=SIZE_AXIS_TEXT),
          axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          axis.title.y = element_text(vjust = 0.6),
          plot.title = element_text(size=SIZE_PLOT_TITLE, 
                                    hjust=0.5),
          legend.title = element_text(colour="black", 
                                      size=SIZE_LEGEND_TITLE), 
          legend.text = element_text(colour="black", 
                                     size=SIZE_LEGEND_TEXT),
          plot.tag = element_text(size = SIZE_LEGEND_TITLE),
          plot.subtitle=element_text(size=SIZE_PLOT_SUBTITLE, 
                                     hjust=0.5))
}




named_palette = c(
  increased_sites = "deeppink3",
  new_sites = "hotpink1",
  decreased_sites = "#2a75b2",
  lost_sites = "lightskyblue",
  astro_nt = "#7FDEEA",
  astro_asyn = "#0097A6",
  neuron_nt = "#FF91C8",
  neuron_asyn = "deeppink3",
  cocult_nt = "#9FA7D8",
  cocult_asyn = "#303F9F"
)


```


```{r load-theme, echo = F, results = "asis"}
## Custom sciRmdTheme with increased width.
## Please ask guillermorocamora@gmail.com for more details.
sciRmdTheme::set.theme(
  theme = "default",
  color = NULL,
  header.sticky = FALSE,
  list.group.icon = "arrow",
  font.family = "Arial",
  font.color = "black",
  header.color = "darkblue"
)

```

```{r}

args <- list(
    # alu_annotations_range_file = here::here("raw_data","
    #                                                  astro_neuron_bulk",
    #                                                    "16_alignment",
    #                                                    "vep",
    #                                                  "alu_annotations_distinct_range_230117.rds"),
    #          reduced_ensemble_range_file = here::here( "raw_data",
    #                                                    "astro_neuron_bulk",
    #                                                    "16_alignment",
    #                                                    "vep",
    #                                                   "reduced_ensembl_38-93_range_230117_stranded_distinct.rds"),
    #          vep_results_range_file = here::here("raw_data",
    #                      "astro_neuron_bulk",
    #                      "16_alignment",
    #                      "vep",
    #                      "vep_results_all_samples_strand_filtered_range.RDS"),
             # gene_expression_file = here::here("raw_data",
             #             "astro_neuron_bulk",
             #             "16_alignment",
             #             "vep",
             #             "astr_cocult_expressed_genes.xlsx"),
             detect_summary_comparison_sig_file = here::here("results",
                                                             "astrocytes",
                                                             "detect_summary_comparison_sig_edits.RDS"),
             # detect_summary_comparison_all_file = here::here("processed_data",
             #                                                 "astro_neuro_coculture",
             #                                                 "detect_summary_comparison_all_edits.RDS"),
             # detect_summary_comparison_alu_file = here::here("processed_data",
             #                                                 "astro_neuro_coculture",
             #                                                 "detect_summary_comparison_sig_alu_edits.RDS"),
             dif_summary_comparison_sig_file = here::here("results",
                                                             "astrocytes",
                                                          "differential_summary_comparison_sig_edits.RDS"),
             dif_summary_comparison_all_file = here::here("results",
                                                             "astrocytes",
                                                             "differential_summary_comparison_all_edits.RDS"),
             # qc_results_dir = here::here("processed_data",
             #                             "astro_neuro_coculture",
             #                             "astro_neuro_editing_QC"),
             # RNA_location_file = here::here("references",
             #                                "All_RNA_subcellular_localization_data.txt"),
             # panel_app_file = here::here("references",
             #                             "panel_app_PD_complex_park_gene_list.tsv"),
             output_folder = here::here("results",
                                        "astrocytes"),
             figure_output_dir = here::here("figures",
                                            "astrocytes")
)
             



plot_labels = c("Decreased_editing" = "Decreased editing",
                "Increased_editing" = "Increased editing",
                "Lost_edit_sites" = "Lost editing sites",
                "New_edit_sites" = "New editing sites",
                "astro_nt_vs_asyn" = "Astrocyte",
                "neuron_nt_vs_asyn" = "Neuron",
                "cocult_nt_vs_asyn" = "Co-culture",
                "astro_nt" = "Astrocyte basal",
                "astro_asyn" = "Astrocyte αsyn-O",
                "cocult_nt" = "Co-culture basal",
                "cocult_asyn" = "Co-culture αsyn-O",
                "neuron_nt" = "Neuron basal",
                "neuron_asyn" = "Neuron αsyn-O",
                "exon" = "Exon",
                "intron" = "Intron",
                "intergenic" = "Intergenic",
                "not_significant" = "Not significant",
                "significant" = "Significant",
                "Astrocyte_increase" = "Astrocyte increase",
                "Astrocyte_decrease" = "Astrocyte decrease",
                "Neuron_increase" = "Neuron increase",
                "Neuron_decrease" = "Neuron decrease",
                "Coculture_increase" = "Co-culture increase",
                "Coculture_decrease" = "Co-culture decrease",
                "no_treatment" = "Basal",
                "1uM_syn_oligomer" = "αsyn-O",
                "increase" = "Increase",
                "decrease" = "Decrease",
                "increased_sites" = "Increased editing rate",
                "decreased_sites" = "Decreased editing rate",
                "treated_edits" = "αsyn-O",
                "untreated_edits" = "Basal",
                "decrease" = "Decreased",
                "increase" = "Increased",
                "new_sites" = "New editing site",
                "lost_sites" = "Lost editing site",
                "significant" = "Significant",
                "not_significant" = "Not significant",
                "astrocyte" = "Astrocyte",
                "neuron" = "Neuron",
                "coculture" = "Co-culture",
                "Astrocyte" = "Astrocyte",
                "Coculture" = "Co-culture",
                "Neuron" = "Neuron"
                )

plot_labeller = as_labeller(plot_labels)
             
```




> Aim: This notebook takes editing data output by jacusa, processes it to add annotations from VEP and repeat-based annotations, and then explores results. It looks at editing results per samples, and also the differential editing results, looking at the difference of each cell line, treated and untreated.
<br><br>



# Methods {.tabset}

## Editing analysis

RNA editing analysis was undertaken with [JACUSA2](https://github.com/dieterich-lab/JACUSA2). This utilises a drichilet multinomial distribution to ascertain whether transcripts are edited at a genomic site in two modes: 

* ‘Detect’ mode compares transcripts against a reference genome identifying editing in individual samples
* ‘Differential’ mode compare two or more samples against each other, looking at a per site level for sites that are differentially edited in one sample compared to another. 


Noting that transcripts with increased editing might not be successfully mapped during the alignment step, multi-sample 2 pass mapping with STAR was rerun, allowing a more generous mismatch rate of 16 base pairs per 100. This did not increase the rate of multimapping during alignment. 


Replicates for each sample group were input to JACUSA2 detect, to output a list of editing sites for each of the six groups: astrocytes untreated, astrocytes treated, coculture untreated, coculture treated, neuron untreated and neuron treated. JACUSA2 differential was used to ascertain those sites that were differentially edited in treated samples of each cell line, compared to untreated. 


All edits were filtered for those that were properly mapped with a pair, and with a z-score greater than 1.96 or less than -1.96. Effects of edits were explored using Variant Effect Predictor (ref).  Repeat annotations were derived from [Repeat Masker 4.0.5](https://www.repeatmasker.org/genomes/hg38/RepeatMasker-rm405-db20140131/hg38.fa.out.gz). Differences in editing sites and rates were explored using R 4.2.0.



# Supplementary code {.tabset}

## Processing Annotation files

The ensembl 38.93 gtf and alu repeat files were pre-processed as per the following script. Note that the alu annotations are unstranded, while the gtf and vep results are stranded.


```{r, eval = F}
source(here::here("scripts", "00a_process_reference_files.R"))

```

### Define levels for filtering

When annotating editing sites with genic/mutation information, the annotations map to the editing sites multiple times. Given this, the multi-annotated editing sites were filtered as follows:

* VEP results are filtered by:
** Whether genes are expressed (as per differential gene expression)
** Then by impact and consequence (highest impact preferenced)
** Then by biotype
** Then by Transcript Support Level (TSL), from lowest (greatest evidence) to highest (least evidence). 
* Repeat annotations were selected in descending order of: Simple_repeat, SINE/Alu, SINE/MIR, LINE/L1, LINE/L2, LTRs, DNAs, Retroposons/SVA, RNAs, Other SINE, Other LINE, Other. If both edits have the same class of repeat annotation, the first is chosen
* Ensembl annotations: Where there are duplicates exonic annotations are chosen first before introns and intergenic. If they have the same genic type, then gene biotype is used to filter, 

<br>
Genic location of the edits were derived from VEP results, with exonic including VEP consequences of start or stop lost, missense, synonymous or coding sequence variant, 3 prime and 5 prime variants and non coding transcript exon variants; intronic including intron variants, splice variants, incomplete terminal codon varians and upstream or downstream variants; and intergenic where specifically ascribed. Where VEP did not assign a variant, genic location was derived from the ensembl 38.93 gtf annotation file, with an exonic region ascribed where any transcript exon lay at the genic location, introns where the location lay within a gene but outside any transcript exon, and otherwise intergenic.


The detailed order for filtering is below:

```{r}
source(here::here("functions_args",
                  "filtering_args.R"))

```



## Editing pipeline

The RNA editing was derived from the raw transcriptomic files using the following workflow:

```{bash run-jacusa, eval = F}

bash /scripts/01a_jacusa_workflow_astrocytes.txt

```


The JACUSA detect results were processed using the following code:
```{r process_jacusa_detect, eval=F}

source(here::here("scripts", "01b_process_detect_astro.R"))
```



The JACUSA differential results were processed using the following code:

```{r process_jacusa_differential, eval = F}
source(here::here("scripts", "01c_process_differential_astro.R"))
```



# Results


## Detect editing data  {.tabset}

```{r}
# Load file of significant results
comparison_df <- readRDS(args$detect_summary_comparison_sig_file)

# Run the following before and after relevelling factors to check no samples lost in processing
na_consequence_detect <- comparison_df %>% 
  dplyr::filter(is.na(Consequence))

# Refactor levels
comparison_df <- comparison_df %>% 
  mutate(Sample = fct_relevel(Sample, filtering_args$sample_levels),
         gene_type = fct_relevel(gene_type, filtering_args$gtf_genic_type),
         Consequence = fct_relevel(Consequence, filtering_args$vep_consequence),
         BIOTYPE = fct_relevel(BIOTYPE, vep_levels_BIOTYPE),
         Consequence_plots = Consequence) # Create additional column with simplified consequences for plotting


detect_export_df <- comparison_df %>% 
    dplyr::select(locus, sample_name, Edits, z.score)
write_csv(detect_export_df,
            file = str_c(args$output_folder, "/sig_edits_astrocyte_detect.csv"))

# Note, when releveling as collapsing levels, the levels have to be defined as a list (not a vector). These are populated from filtering_args, though, being defined as a list means that they get directly loaded to the environment, not as a subset of filtering_args).

levels(comparison_df$Consequence_plots) <- vep_consequence_plots
levels(comparison_df$BIOTYPE) <- vep_levels_BIOTYPE

```



### Editing site numbers across cultures and genic regions


In the untreated samples, there were 78,547 editing sites in the astrocyte monoculture, 103,340 in the neuron monoculture, and 100,740 in the coculture. With a-syn treatment, there were at least an 11% increase in editing sites in each sample: 15.9% in astrocytes (p value of exonic proportion change measured by chi squared test 2e-16), 26.2% in neurons (p = 2e-16), and 11.1% in cocultures (p = 2e-16). In astrocyte containing samples, 50.3 - 62.4% of editing was seen in exonic regions, while in the neuronal line most editing was seen in intronic regions: 57.8 and 61.4% in untreated and treated samples respectively. 


Summary of editing site number (percent) across cultures and genic locations.

```{r, results = "asis", fig.cap = "Table showing editing site numbers (percent) for the different cultures across genic locations" }
summary_numbers <- summary_by_variable(comparison_df,
                    sample_metric = Sample)[[2]] 


genic_summary <- summary_by_variable_for_publication(comparison_df,
                    sample_metric = "Sample",
                    variable = "gene_type")[[2]] %>% 
    left_join(as_tibble(summary_numbers),
              by = "Sample") %>% 
    dplyr::rename(Total = n) 

write_csv(genic_summary,
          file = str_c(args$output_folder, "/editing_site_numbers_genicsite_summary.csv"))


genic_summary %>% 
    as_tibble() %>% 
      DT::datatable(rownames = FALSE,
                     options = list(scrollX = TRUE))

```

```{r}

# Extract values for _nt and _asyn conditions for each group
astro_nt <- summary_numbers["astro_nt"]
astro_asyn <- summary_numbers["astro_asyn"]
cocult_nt <- summary_numbers["cocult_nt"]
cocult_asyn <- summary_numbers["cocult_asyn"]
neuron_nt <- summary_numbers["neuron_nt"]
neuron_asyn <- summary_numbers["neuron_asyn"]

# Calculate percent increase for each group
percent_increase_astro <- round(((astro_asyn - astro_nt) / astro_nt) * 100, 1)
percent_increase_cocult <- round(((cocult_asyn - cocult_nt) / cocult_nt) * 100, 1)
percent_increase_neuron <- round(((neuron_asyn - neuron_nt) / neuron_nt) * 100, 1)

# Print the results
cat("Percent increase in editing sites for astrocyte cultures:", percent_increase_astro, "%\n")
cat("Percent increase in editing sites for cocultures:", percent_increase_cocult, "%\n")
cat("Percent increase in editing sites for neuron cultures:", percent_increase_neuron, "%\n")
```





```{r}

detect_genic_plot <- compare_plot(comparison_df %>%
               drop_na(gene_type),
             proportional = "stack",
             comparison = (gene_type),
             plot_label = plot_labels) +
    labs(y = "Number of editing sites",
         fill = "Genic location") +
  theme_plot()


ggsave(detect_genic_plot,
       file = "detect_genic_plot.pdf",
       path = args$figure_output_dir,
       device = "pdf",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)

ggsave(detect_genic_plot,
       file = "detect_genic_plot.png",
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)

detect_genic_plot +
  labs(title = "Number of editing sites by genic location") +
       theme_aw

```




#### Check chi square test of independence of categorical variables


```{r}
print("Astrocytes: significant change in exonic ratio")
chi_square_factor_test(comparison_df,
                       culture_type = "astrocyte",
                       variable_of_interest = gene_type, # The variable does not get ""
                       factor_of_interest = "exon")

print("Coculture: significant change in exonic ratio")
chi_square_factor_test(dataframe = comparison_df,
                            culture_type = "coculture",
                            variable_of_interest = gene_type, # The variable does not get ""
                       factor_of_interest = "exon")


print("Neuron: significant change in exonic ratio")
chi_square_factor_test(dataframe = comparison_df,
                            culture_type = "neuron",
                            variable_of_interest = gene_type, # The variable does not get ""
                       factor_of_interest = "exon")
```



<br>


### Editing by Consequence
```{r, fig.width = 20, results = "asis"}

summary_by_variable_for_publication(comparison_df,
                                               sample_metric = "Sample",
                                               variable = "Consequence_plots")[[1]] %>% 
      DT::datatable(rownames = FALSE,
                     options = list(scrollX = TRUE))

```


```{r}

compare_plot(comparison_df %>%
               drop_na(Consequence_plots),
                    proportional = "fill",
                    comparison = (Consequence_plots),
             plot_label = plot_labels) +
    labs(title = "Consequence of edits", y = "Proportion of editing sites", fill = "Consequence",
         subtitle = "Excluding NAs") +
  theme_plot() +
       theme_aw


```


<br>


### Editing by Biotype
```{r, fig.width = 20, results = "asis"}

summary_by_variable_for_publication(comparison_df,
                                               sample_metric = "Sample",
                                               variable = "BIOTYPE")[[1]] %>% 
      DT::datatable(rownames = FALSE,
                     options = list(scrollX = TRUE))


```


```{r}
compare_plot(comparison_df %>%
               drop_na(BIOTYPE),
                    proportional = "fill",
                    comparison = BIOTYPE,
             plot_label = plot_labels) +
    labs(title = "Biotype of edits", y = "Proportion of editing sites",
         fill = "Transcript biotype") +
  guides(fill = guide_legend(reverse = TRUE)) +
  theme_plot() +
       theme_aw

```
<br>

### Editing by repeat regions

```{r}
summary_by_variable_for_publication(comparison_df,
                                               sample_metric = "Sample",
                                               variable = "repeat_class")[[1]] %>% 
      DT::datatable(rownames = FALSE,
                     options = list(scrollX = TRUE))
```





```{r}
# Note: NAs have been removed
repeat_types_plot <- compare_plot(comparison_df %>%
                                           drop_na(repeat_class),
                    proportional = "fill",
                    comparison = fct_rev(repeat_class),
                    plot_label = plot_labels) +
    labs( y = "Proportion of editing sites", fill = "Repeat class") +
    guides(fill = guide_legend(reverse = TRUE)) +
  theme_plot()


ggsave(repeat_types_plot,
       file = "repeat_types_plot.pdf",
       path = args$figure_output_dir,
       device = "pdf",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)

ggsave(repeat_types_plot,
       file = "repeat_types_plot.png",
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)

repeat_types_plot +
  labs(title = "Proportion of repeat types",
       subtitle = "Excluding NAs") +
    theme_aw


```

<br>

### Comparison of editing rates at a given site between samples
```{r}
comparison_df %>%
  group_by(Sample) %>%
  summarise(min_Edits = min(Edits, na.rm = T),
            median_Edits = median(Edits, na.rm =T),
            IQR_Edits = IQR(Edits, na.rm = T),
            max_Edits = max(Edits, na.rm = T)
           )  %>% 
      DT::datatable(rownames = FALSE,
                     options = list(scrollX = TRUE))
```


<br>
<br>



## Differential editing data {.tabset}

### Summary of differentially edited sites

```{r}
annotated_dif_data  <- readRDS(args$dif_summary_comparison_all_file)

# Add columns of interest
annotated_dif_data <- annotated_dif_data %>%
  mutate(treat_untreat_dif = treated_edits - untreated_edits) %>%
  dplyr::relocate(treat_untreat_dif, .after=treat_untreat_ratio) %>%
  mutate(culture = str_replace_all(SampleName, c("astro_nt_vs_asyn" = "Astrocyte",
                                                "cocult_nt_vs_asyn" = "Coculture",
                                                "neuron_nt_vs_asyn" = "Neuron")) %>%
           as.factor,
         binary_editing_change = ifelse(treat_untreat_dif < 0.00, "decrease",
                                      ifelse(treat_untreat_dif > 0.00, "increase",
                                             "Other"
                                             )
                                 ) %>%
           as.factor(),
         four_editing_changes = case_when(
           treat_untreat_ratio == Inf ~ "new_sites",
           treat_untreat_ratio != Inf & treat_untreat_dif > 0 ~ "increased_sites",
           treat_untreat_ratio != 0 & treat_untreat_dif < 0 ~ "decreased_sites",
           treat_untreat_ratio == 0 ~ "lost_sites") %>%
           as.factor(),
         Sample_change = str_c(culture, "_", binary_editing_change) %>%
           as.factor(),
         Consequence_plots = Consequence,  # Add in consequence row with simplified levels for plotting
         sig_edit = as_factor(sig_edit))

# Rationalise levels in factors
levels(annotated_dif_data$Consequence_plots) <- vep_consequence_plots
levels(annotated_dif_data$BIOTYPE) <- vep_levels_BIOTYPE
annotated_dif_data$four_editing_changes <- factor(annotated_dif_data$four_editing_changes, levels = c("new_sites", "increased_sites", "decreased_sites", "lost_sites"))

# Filter for significance
sig_annotated_dif_data <- annotated_dif_data %>%
  dplyr::filter(sig_edit == "significant")

differential_export_df <- sig_annotated_dif_data %>%
    dplyr::select(locus, SampleName, untreated_edits, treated_edits, z.score)
write_csv(differential_export_df,
            file = str_c(args$output_folder, "/sig_edits_astrocyte_differential.csv"))



```



```{r}
# Creat summary table
sig_annotated_dif_table <- sig_annotated_dif_data %>%
    dplyr::select(culture, four_editing_changes) %>%
    table() %>% 
    as_tibble() %>% 
     spread(four_editing_changes, n) 

sig_annotated_dif_table %>% 
        DT::datatable(rownames = FALSE,
                  options = list(scrollX = TRUE))

```






```{r}
dif_site_summary_plot <- sig_annotated_dif_data %>%
  dplyr::mutate(four_editing_changes = fct_relevel(four_editing_changes, levels = c("new_sites", "increased_sites", "lost_sites", "decreased_sites"))) %>%
  ggplot(aes(x = binary_editing_change, fill = four_editing_changes)) +
  geom_bar(position = "stack") +
  facet_wrap(vars(culture),
             ncol = 3,
             labeller = plot_labeller) +
  scale_fill_manual(values = named_palette,
                    labels = plot_labels) +
  theme_plot() +
  labs(x = "Change in editing per site",
       y = "Number of editing sites",
       fill = element_blank()) +
   scale_x_discrete(labels = element_blank())


ggsave(dif_site_summary_plot,
       file = "differential_site_summary_plot.pdf",
       path = args$figure_output_dir,
       device = "pdf",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)


ggsave(dif_site_summary_plot,
       file = "differential_site_summary_plot.png",
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)

dif_site_summary_plot +
  labs(title = "Changes in differentially edited sites")

```



```{r, eval=F}
# Export for cluster profiler
sum_sig_dif_data <- sig_annotated_dif_data %>%
  dplyr::select(locus, Sample_change, culture, Gene, SYMBOL, treat_untreat_dif, treat_untreat_ratio, z.score, gene_type, Consequence, Consequence_plots, BIOTYPE, repeat_class, ARE_binary, ARE_cluster) %>%
  dplyr::mutate(log_ratio = log2(treat_untreat_ratio)) %>%
  dplyr::filter(gene_type != "intergenic") %>%  # select only those sites within genes
  arrange(culture, desc(log_ratio))

write_csv(sum_sig_dif_data,
          str_c(args$output_folder,
                "/summary_results_genes.csv"))


sum_all_dif_data <- annotated_dif_data %>%
  dplyr::select(locus, Sample_change, culture, Gene, SYMBOL, treat_untreat_dif, treat_untreat_ratio, z.score, gene_type, Consequence, Consequence_plots, BIOTYPE, repeat_class, ARE_binary, ARE_cluster) %>%
  dplyr::mutate(log_ratio = log2(treat_untreat_ratio)) %>%
  dplyr::filter(gene_type != "intergenic") %>%
  arrange(culture, desc(log_ratio))

sum_all_dif_data
write_csv(sum_all_dif_data,
          str_c(args$output_folder,
                "/dge_comparisons",
                "/summary_results_all_genes_inc_nonsig.csv"))

all_gene_background <- sum_all_dif_data %>%
  dplyr::select(Gene, SYMBOL) %>%
  distinct(Gene, .keep_all = T)

write_csv(all_gene_background,
          str_c(args$output_folder,
                "/dge_comparisons",
                "/summary_results_all_dif_genelist_background_universe.csv"))


sig_increase_gene_ratio <- sum_sig_dif_data %>%
  dplyr::filter(treat_untreat_dif > 0,
                Consequence_plots == "3 prime UTR variant") %>%
  distinct(SYMBOL, .keep_all = TRUE)

write_csv(sig_increase_gene_ratio,
          str_c(args$output_folder,
                "/dge_comparisons",
                "/significant_3UTR_genes_ratios.csv"))

sig_increase_gene_difference <- sum_sig_dif_data %>%
  dplyr::filter(treat_untreat_dif > 0,
                Consequence_plots == "3 prime UTR variant") %>%
  arrange(culture, desc(treat_untreat_dif)) %>%
  distinct(SYMBOL, .keep_all = TRUE)
sig_increase_gene_difference

write_csv(sig_increase_gene_difference,
          str_c(args$output_folder,
                "/dge_comparisons",
                "/significant_3UTR_genes_difference.csv"))


```


<br>

### Differential editing rate at a given site


```{r, fig.height=4, fig.width=10}
par_coord_plot_data <- sig_annotated_dif_data %>%
  dplyr::select(locus, culture, four_editing_changes, Sample_change, untreated_edits, treated_edits)  %>%
   pivot_longer(cols = c("untreated_edits", "treated_edits"), # Pivot longer for ggplot
               names_to = "timepoint",
               values_to = "editing_rate") %>%
  dplyr::filter(four_editing_changes %in% c("increased_sites","decreased_sites")) %>%
  dplyr::mutate(four_editing_changes = factor(four_editing_changes, levels = c("increased_sites","decreased_sites")),
                timepoint = factor(timepoint, levels = c("untreated_edits", "treated_edits"))
                )

parallel_line_plot <- par_coord_plot_data %>%
  arrange(four_editing_changes, locus)  %>%
  dplyr::mutate(locus = forcats::fct_reorder(as_factor(locus), four_editing_changes, .fun=first)) %>% #Reorder locus as a factor based on four editing changes so it will plot correctly
  slice(1:nrow(par_coord_plot_data)) %>%
ggplot(aes(x = timepoint, y = editing_rate, group = locus, colour = four_editing_changes)) +
   geom_path(alpha = 0.1, linewidth = 0.1) +
  theme_plot() +
  facet_wrap(~culture,
             labeller = plot_labeller) +
  scale_colour_manual(values = named_palette,
                        labels = plot_labels) +
  scale_x_discrete(labels = plot_labels) +
  labs(x = element_blank(), y = "Editing rate", colour = "Editing rate change") +
  guides(colour = guide_legend(override.aes = list(alpha = 1, linewidth = 1))) # Change legend so can see the colours

ggsave(parallel_line_plot,
       file = "differential_editing_parallel_line_plot_tall.pdf",
       path = args$figure_output_dir,
       device = "pdf",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)


ggsave(parallel_line_plot,
       file = "differential_editing_parallel_line_plot_tall.png",
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)

parallel_line_plot +
  labs(title = "Change in editing rate with a-syn treatment") +
       theme_aw

```




<br>

### Proportions of differential editing: Site consequence

```{r}

# For the plot of consequences, row bind the baseline (untreated) VEP Consequences for each cell line, with those that change in differential editing to look at difference.

consequence_plot_data <- comparison_df %>%
  dplyr::filter(treatment == "no_treatment") %>%
  dplyr::select(locus, culture, binary_editing_change = treatment, Consequence_plots) %>%
  dplyr::mutate(culture = str_to_title(culture)) %>% #Make culture column upper case to match the differential data
  bind_rows(
    sig_annotated_dif_data %>%
              dplyr::select(locus, culture, binary_editing_change, Consequence_plots)
  )



consequence_dif_plot <- consequence_plot_data %>%
  drop_na(Consequence_plots) %>%
  ggplot(aes(x=binary_editing_change, fill = fct_rev(Consequence_plots))) +
  geom_bar(position = "fill") +
    scale_fill_paletteer_d("ggthemes::Classic_20",
                           na.value = "grey90") +
    theme_plot() +
    labs(y = "Proportion of editing sites", x = NULL, fill = "Consequence") +
   scale_x_discrete(labels = plot_labels)  +
  guides(fill = guide_legend(reverse = TRUE)) +
  facet_wrap(vars(culture),
             labeller = plot_labeller)





ggsave(consequence_dif_plot,
       file = "differential_consequence_plot.pdf",
       path = args$figure_output_dir,
       device = "pdf",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)

ggsave(consequence_dif_plot,
       file = "differential_consequence_plot.png",
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)


consequence_dif_plot +
          labs(title = "Consequences of differential editing changes",
         subtitle = "Baseline verse sites with decreased/increased editing with treatment") +
       theme_aw


```




Check chi square: proportion of sites in 3UTR in increased verse decreased sites
```{r}

print("Astrocytes")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Astrocyte",
                                    variable_of_interest = Consequence_plots,
                                    factor_of_interest = '3 prime UTR variant' )

print("Neuron")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Neuron",
                                    variable_of_interest = Consequence_plots,
                                    factor_of_interest = '3 prime UTR variant' )

print("Cocultures")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Coculture",
                                    variable_of_interest = Consequence_plots,
                                    factor_of_interest = '3 prime UTR variant' )

```


<br>

### Proportions of differential editing: Site biotype


```{r}
# Biotype with baseline
biotype_plot_data <- comparison_df %>%
  dplyr::filter(treatment == "no_treatment") %>%
  dplyr::select(locus, culture, binary_editing_change = treatment, BIOTYPE) %>%
  dplyr::mutate(culture = str_to_title(culture)) %>% #Make culture column upper case to match the differential data
  bind_rows(
    sig_annotated_dif_data %>%
              dplyr::select(locus, culture, binary_editing_change, BIOTYPE)
  )

biotype_dif_plot <- biotype_plot_data %>%
  drop_na(BIOTYPE) %>%
  ggplot(aes(x=binary_editing_change, fill = fct_rev(BIOTYPE))) +
  geom_bar(position = "fill") +
    scale_fill_paletteer_d("ggthemes::Classic_20",
                           na.value = "grey90") +
    theme_plot() +
    labs(y = "Number of editing sites", x = NULL, fill = "Biotypes") +
   scale_x_discrete(labels = plot_labels)  +
  guides(fill = guide_legend(reverse = TRUE)) +
  facet_wrap(vars(culture),
             labeller = plot_labeller)

biotype_dif_plot +
          labs(title = "Proportion of biotypes of differential editing changes",
         subtitle = "Baseline verse sites with decreased/increased editing with treatment") +
       theme_aw

ggsave(biotype_dif_plot,
       file = "supp_differential_biotype_plot.png",
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)



```


```{r}
 
print("Astrocytes")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Astrocyte",
                                    variable_of_interest = BIOTYPE,
                                    factor_of_interest = 'Protein coding' )

print("Coculture")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Coculture",
                                    variable_of_interest = BIOTYPE,
                                    factor_of_interest = 'Protein coding' )

print("Neuron")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Neuron",
                                    variable_of_interest = BIOTYPE,
                                    factor_of_interest = 'Protein coding' )



```


<br>

### Proportions of differential editing: Alu sites



```{r}
# Biotype with baseline
alu_plot_data <- comparison_df %>%
  dplyr::filter(treatment == "no_treatment") %>%
  dplyr::select(locus, culture, binary_editing_change = treatment, repeat_class) %>%
  dplyr::mutate(culture = str_to_title(culture)) %>% #Make culture column upper case to match the differential data
  bind_rows(
    sig_annotated_dif_data %>%
              dplyr::select(locus, culture, binary_editing_change, repeat_class)
  )

alu_dif_plot <- alu_plot_data %>%
  drop_na(repeat_class) %>%
  ggplot(aes(x=binary_editing_change, fill = (repeat_class))) +
  geom_bar(position = "fill") +
    scale_fill_paletteer_d("ggthemes::Classic_20",
                           na.value = "grey90") +
    theme_plot() +
    labs(y = "Number of editing sites", x = NULL, fill = "Repeat class") +
   scale_x_discrete(labels = plot_labels)  +
  guides(fill = guide_legend(reverse = TRUE)) +
  facet_wrap(vars(culture),
             labeller = plot_labeller)

alu_dif_plot  +
          labs(title = "Proportion of repeat classes of differential editing changes",
         subtitle = "Baseline verse sites with decreased/increased editing with treatment") +
       theme_aw

ggsave(alu_dif_plot,
       file = "supp_differential_repeat_plot.png",
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 240,
       height = 180,
       limitsize = TRUE
)



```


Testing proportional changes for SINE/Alu
```{r}
 
print("Astrocytes")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Astrocyte",
                                    variable_of_interest = repeat_class,
                                    factor_of_interest = "SINE/Alu" )

print("Coculture")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Coculture",
                                    variable_of_interest = repeat_class,
                                    factor_of_interest = "SINE/Alu"  )

print("Neuron")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Neuron",
                                    variable_of_interest = repeat_class,
                                    factor_of_interest = "SINE/Alu"  )



```



Testing proportional changes for Simple_repeat
```{r}
 
print("Astrocytes")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Astrocyte",
                                    variable_of_interest = repeat_class,
                                    factor_of_interest = "Simple_repeat" )

print("Coculture")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Coculture",
                                    variable_of_interest = repeat_class,
                                    factor_of_interest = "Simple_repeat"  )

print("Neuron")
chi_square_factor_test_differential(dataframe = sig_annotated_dif_data,
                                    culture_type = "Neuron",
                                    variable_of_interest = repeat_class,
                                    factor_of_interest = "Simple_repeat" )



```

<br
<br>

# Combine plots 



```{r, fig.width = 20, fig.height = 18, units = "cm"}
summary_edit_site_cons_plot <- (detect_genic_plot + dif_site_summary_plot) /
  parallel_line_plot /
  (consequence_dif_plot + plot_spacer()) +
  plot_annotation(tag_levels = 'a')

ggsave(summary_edit_site_cons_plot  ,
       file = "summary_edit_site_cons_plot.pdf",
       path = args$figure_output_dir,
       device = "pdf",
       dpi = 300,
       units = "mm",
       width = 380,
       height = 450,
       limitsize = TRUE
)

ggsave(summary_edit_site_cons_plot  ,
       file = "summary_edit_site_cons_plot.png",
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 380,
       height = 450,
       limitsize = TRUE
)

summary_edit_site_cons_plot +
       theme_aw

```


```{r, include = F, fig.width = 16, fig.height = 16, units = "cm"}

edit_motif_summary_plot <-
    consequence_dif_plot /
    biotype_dif_plot /
    alu_dif_plot +
    plot_annotation(tag_levels = 'a')

edit_motif_summary_plot +
       theme_aw

ggsave(edit_motif_summary_plot ,
       file = "summary_edit_motif_plot.pdf",
       path = args$figure_output_dir,
       device = "pdf",
       dpi = 300,
       units = "mm",
       width = 260,
       height = 240,
       limitsize = TRUE
)

ggsave(edit_motif_summary_plot ,
       file = "summary_edit_motif_plot.png",
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 260,
       height = 240,
       limitsize = TRUE
)


```




```{r, include = F, fig.width = 20, fig.height = 8}
dif_summary_plot <- (dif_site_summary_plot + parallel_line_plot) +
  plot_layout(widths = c(1, 1.5)) +
       theme_aw

ggsave(dif_summary_plot,
       file = "dif_summary_plot.png",
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 450,
       height = 160,
       limitsize = TRUE
)
```


<!-- # Conclusions -->
<!-- What are the conclusions to be drawn from this analysis? -->

<!-- # Next steps -->
<!-- What are the next steps after this? -->

# Session info

<details>
  <summary>Show/hide</summary>

```{r reproducibility, echo = FALSE}
# Session info
library("sessioninfo")
options(width = 120)
session_info()
```

</details>
