---
title: "acta_editing_analysis_3_disease_w_gene_summary"
author: 
- name: "Aaron Wagen"
  affiliation: UCL
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  bookdown::html_document2:
    figure_caption: yes
    code_folding: hide
    theme: paper
    highlight: haddock
    df_print: paged
    toc: true
    toc depth: 3
    toc_float: true
    number_sections: true
  md_document:
    variant: markdown_github
    toc: true
    number_sections: true
always_allow_html: true
link-citations: true
notes-after-punctuation: false
---


```{r setup, include = FALSE}
library(here) # For project-specific paths
library(tidyverse)
library(GenomicRanges)
library(plyranges)
library(paletteer)
library(patchwork)
library(paletteer)
library(ggtranscript)
library(GGally)
library(clusterProfiler)
library(imager)
library(broom)
library(ggforce)
library(tidycat)
library(performance)
library(MASS)
library(data.table)
library(EWCE)


knitr::opts_chunk$set(echo = T, warning = F, message = F)
options(bitmapType='cairo')
# Set defaults for ggplots 
theme_aw <- theme_bw(base_family = "Helvetica",
           base_size = 12) + 
  theme(panel.grid.major.x = element_blank(),
        legend.position = "right",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.y = element_text(vjust = 0.6),
        panel.spacing = unit(0.1, "lines"))

here::i_am("docs/acta_editing_analysis_3_disease_w_gene_summary.Rmd")

# Load functions
source(here::here("r_scripts",
                  "r_functions.R"))

source(here::here("r_scripts",
                  "final_editing_functions_editing_summary_metric.R"))


source(here::here("r_scripts",
                  "filtering_args.R"))


CORES=1

palette = c("#2a75b2", "grey75", "#B0C6E7", "deeppink3")
#palette = c("cyan4", "grey75")

## Set options
options(dplyr.summarise.inform = FALSE)
options(lifecycle_verbosity = "warning")
knitr::opts_chunk$set(echo = T, warning = F, message = F, out.width="100%", fig.align = "center", dpi = 100)
```





```{r load-theme, echo = F, results = "asis"}
## Custom sciRmdTheme with increased width.
## Please ask guillermorocamora@gmail.com for more details.
sciRmdTheme::set.theme(
  theme = "default",
  color = NULL,
  header.sticky = FALSE,
  list.group.icon = "arrow",
  font.family = "Arial",
  font.color = "black",
  header.color = "darkblue"
)


## Custom styles for the output html
cat('
<style type="text/css">
.dataTables_scrollHeadInner{
  width:100% !important;
}
.dataTables_scrollHeadInner table{
  width:100% !important;
}
/*
.code-folding-btn {
  display: none;
}
*/
h3, .h3 {
  font-size: 22px!important;
}
h4, .h4 {
  font-size: 18px!important;
}
h5, .h5 {
  font-size: 16px!important;
}
body{
  font-size: 13px;
}
pre {
  line-height: 1;
}
.tocify-subheader {
    text-indent: 15px;
    display: none;
    font-size: 12px;
}
</style>')


```


```{r}
args <- list(
   detect_summary_comparison_sig_file = here::here("results",
                                                  "brain",
                                                  "jacusa_detect",
                                                  "detect_summary_comparison_sig_edits.RDS"),
   # detect_summary_folder = here::here("results",
   #                                     "brain",
   #                                     "jacusa_detect"),
    sample_annotation_file = here::here("raw_data",
                                        "acta_neuropath_bulk",
                                        "SampleAnnot_metadata.txt"),
    output_folder = here::here("results"),
  lengths_summary_path = here::here("references", 
                                    "ensembl93_gene_lengths_summary.RDS"),
  reduced_gtf = here::here("references", "ensembl93_reduced_gtf.RDS")
)


# Args for trial
# args <- list(
#    detect_summary_comparison_sig_file =  "/nemo/lab/gandhis/home/users/wagena/Astrocytes_editing/results/brain/jacusa_detect/detect_summary_comparison_sig_edits.RDS",
#    sample_annotation_file = "/nemo/lab/gandhis/home/users/wagena/AIediting/raw_data/acta_neuropath_bulk/SampleAnnot_metadata.txt"
# )



plot_labels = c("Decreased_editing" = "Decreased editing",
                "Increased_editing" = "Increased editing",
                "Lost_edit_sites" = "Lost editing sites",
                "New_edit_sites" = "New editing sites",
                "astro_nt_vs_asyn" = "Astrocyte untreated vs a-syn",
                "neuron_nt_vs_asyn" = "Neuron untreated vs a-syn",
                "cocult_nt_vs_asyn" = "Coculture untreated vs a-syn",
                "astro_nt" = "Astrocyte untreated",
                "astro_asyn" = "Astrocyte a-syn",
                "cocult_nt" = "Coculture untreated",
                "cocult_asyn" = "Coculture a-syn",
                "neuron_nt" = "Neuron untreated",
                "neuron_asyn" = "Neuron a-syn",
                "exon" = "Exon",
                "intron" = "Intron",
                "intergenic" = "Intergenic",
                "not_significant" = "Not significant",
                "significant" = "Significant",
                "Astrocyte_increase" = "Astrocyte increase",
                "Astrocyte_decrease" = "Astrocyte decrease",
                "Neuron_increase" = "Neuron increase",
                "Neuron_decrease" = "Neuron decrease",
                "Coculture_increase" = "Coculture increase",
                "Coculture_decrease" = "Coculture decrease",
                "no_treatment" = "Baseline",
                "increase" = "Increase",
                "decrease" = "Decrease",
                "increased_sites" = "Increased editing rate",
                "decreased_sites" = "Decreased editing rate",
                "treated_edits" = "a-syn",
                "untreated_edits" = "Untreated",
                "exon" = "Exon",
                "intron" = "Intron",
                "intergenic" = "Intergenic",
                "increased" = "Increase",
                "decreased" = "Decreased",
                "Disease_Group" = "Disease Group",
                ".resid" = "Corrected residuals"
                )

plot_labeller = as_labeller(plot_labels)



named_palette = c(
  increased_sites = "deeppink3",
  increased = "deeppink3",
  new_sites = "hotpink1",
  decreased_sites = "#2a75b2",
  decreased = "#2a75b2",
  lost_sites = "lightskyblue",
  astro_nt = "#7FDEEA",
  astro_asyn = "#0097A6",
  neuron_nt = "#FF91C8",
  neuron_asyn = "deeppink3",
  cocult_nt = "#9FA7D8",
  cocult_asyn = "#303F9F",
  Control = "#B0BFBB",
  no_treatment = "#1d9064",
  DLB = "#ce490a",
  PD  = "#187853",
  `1uM_syn_oligomer`= "#625aa4",
  PDD = "#de0077",
    exon = "#fc6a0f",
  intron = "#9db8e0",
  intergenic = "#1a62a5"
)

```


## Background data re samples

Sample annotation and information

```{r}
# Load gene lengths
gene_lengths <- readRDS(args$lengths_summary_path)

editing_metrics <- c("Edits", "edit_adjusted")
outcome_to_explore = "Disease_Group"
covariates_used = c("RIN", "Sex")

```


```{r}
# Load annotation file
metadata <- read_delim(args$sample_annotation_file)
metadata <- metadata %>% 
   dplyr::select(Sample = sample_name, Disease_Group, Sex, AoO, AoD, DD, PMI, RIN = RINe_bulkRNA_Tapestation, aSN, TAU, AB = `thal AB`, aCG_aSN_score = `aCG aSN score`) %>% 
  dplyr::mutate(sample_id = Sample) %>% 
    dplyr::filter(Disease_Group %in% c("PD", "Control"))


```


Samples per disease group


```{r, fig.width = 16, fig.height = 7}
RIN_bar <- metadata %>% 
  ggplot(aes(x=Sample, y = RIN, fill = Disease_Group)) +
  geom_bar(stat = "identity") +
  theme_aw +
    scale_fill_manual(values = named_palette) + 
  labs(title = "Samples RIN")
  

AOD_bar <- metadata %>% 
  ggplot(aes(x=Sample, y = AoD, fill = Disease_Group)) +
  geom_bar(stat = "identity") +
  theme_aw +
      scale_fill_manual(values = named_palette) + 
  labs(title = "Age of death")

PMI_bar <- metadata %>% 
  ggplot(aes(x=Sample, y = PMI, fill = Disease_Group)) +
  geom_bar(stat = "identity") +
  theme_aw +
      scale_fill_manual(values = named_palette) + 
  labs(title = "Post mortem interval")

sex_bar <- metadata %>% 
  ggplot(aes(x=Disease_Group, fill = Sex)) +
  geom_bar(stat = "count",
           position = "stack") +
  theme_aw +
      scale_fill_paletteer_d("werpals::pan",
                        direction = -1) +
  labs(title = "Samples: sex")



(RIN_bar | AOD_bar | PMI_bar | sex_bar | plot_layout(guides = "collect", widths = c(2,2,2,1)))

```


```{r}
# Load file of significant results
# master_df <- readRDS(args$detect_summary_comparison_sig_file)
# master_df$Disease_Group <- factor(master_df$Disease_Group)
```

# Load edit site files

```{r}
# Load file of significant results
master_df <- readRDS(args$detect_summary_comparison_sig_file) 
# Replace NA gene types with intergenic (in efficient manner)
master_df$gene_type[is.na(master_df$gene_type)] <- "intergenic"
master_df$Disease_Group <- factor(master_df$Disease_Group)


```



```{r}
#contrasts = list(PD = c("Control", "PD"))
contrast_to_explore <- c("Control", "PD")
region_to_explore <- NULL
```



# Results {.tabset}

## Text

In order to fulfil linearity assumptions, editing results were filtered to include those >0 and <1. Linear regression was utilised to explore relationships between PD and control brains, correcting for the covariates of RIN and sex.


```{r,  units = "cm", fig.width = 26, fig.height = 22}
# Analysis and plotting code for the current region and contrast
    contrast_list <- list(Control = contrast_to_explore[1],
                          Disease = contrast_to_explore[2])
    disease = contrast_to_explore[2]
    
    comparison_df <- master_df %>% 
        dplyr::filter(Disease_Group %in% contrast_to_explore)
    
    # Write output table
    summary_edits_df <- comparison_df %>% 
        dplyr::select(seqnames, start, end, strand, Sample, Edits, z.score)
    write_delim(summary_edits_df,
               file = str_c(args$output_folder, "/brain/sig_edits_", contrast_to_explore[1], "_", contrast_to_explore[2], ".tsv"))
    
    # Replace NAs in genic type with intergenic, and create column of collapsed biotypes for plotting
   comparison_df <- comparison_df %>%
      dplyr::mutate(biotype_plots = as_factor(gene_biotype)) %>%
      remove_suffix_from_gene_id(., column = "gene_id")
    
    # Recode and reorder biotype levels for plotting
    levels(comparison_df$biotype_plots) <- vep_levels_BIOTYPE

    # Factorise locus
    comparison_df$locus <- factor(comparison_df$locus)
    
    # Factorise disease group, checking levels in data match levels in factor to be
         comparison_df$Disease_Group <- factor(comparison_df$Disease_Group)
    dataframe_levels <- levels(comparison_df$Disease_Group)
    if (!all(dataframe_levels %in% contrast_to_explore) && all(contrast_to_explore %in% dataframe_levels)) {
      errorCondition("Supplied sorting factor does not match disease group levels in data")
    }
    levels(comparison_df$Disease_Group) <- contrast_to_explore
    
    # Find sample counts and consequences for annotation later
    sample_counts <- comparison_df %>%
        ungroup() %>% 
      distinct(Sample, Disease_Group) %>%
      count(Disease_Group)
    
    loci_consequences <- comparison_df %>%
      dplyr::select(locus, gene_name, gene_id, biotype_plots, gene_type) %>%
      dplyr::distinct(locus, .keep_all = T)
    
    # Filter by number of sites common within groups and between samples
    sites_to_explore <- data.table::copy(comparison_df)
    sites_to_explore <- sites_to_explore %>% 
           dplyr::filter(Edits < 1)   #Filter to editing rate < 1.
    number_common_samples_per_group = 2
    site_label <- "common_2_sample"
    site_annotation <- "Sites common to 2 samples per group: "
    
    common_loci <- find_sites_loci_common_to_x_samples(sites_df = sites_to_explore,
                                                       sample_counts_df = sample_counts,
                                                       no_common_samples_per_site = number_common_samples_per_group )
    
    sites_subset <- filter_sites_by_locus_vector(sites_df = sites_to_explore,
                                                 vector_of_loci = common_loci,
                                                 factor_list = contrast_list)
    
    
    number_sites_per_sample <- sites_subset %>% 
      count(Sample) %>% 
      dplyr::rename(n_sites_per_sample = n)
    
    sites_subset <- sites_subset %>% 
      left_join(number_sites_per_sample,
                by = "Sample")
    
    
    ###  Regression by sites #### 
    lm_results_sites <- regress_editing(sites_df = sites_subset,
                                        editing_metric = "Edits",
                                        covariates = covariates_used,
                                        factor_list = contrast_list,
                                        outcome_measure = outcome_to_explore,
                                        disease = disease)
    
    summary(lm_results_sites$lm_results_with_outcome)
    tidy(lm_results_sites$lm_results_with_outcome, conf.int = T)
```


A sensitivity analysis was performed additionally covarying for the number of sites per sample.


```{r,  units = "cm", fig.width = 10, fig.height = 8}
### Run regression covaried by n sites per sample
    # Regression by number of sites + n_sites_per_sample
    lm_results_sites_nsites <- regress_editing(sites_df = sites_subset,
                                               editing_metric = "Edits",
                                               covariates = c("n_sites_per_sample", covariates_used),
                                               factor_list = contrast_list,
                                               outcome_measure = outcome_to_explore,
                                               disease = disease)

summary(lm_results_sites_nsites$lm_results_with_outcome)
tidy(lm_results_sites_nsites$lm_results_with_outcome, conf.int = T)
```





## Figures





```{r}
total_sites_violin <- plot_site_number_violin(sites_subset, plot_label = plot_labels)

# Violin plot
 edit_resid_violin <-  lm_results_sites$sites_subset_residuals %>% 
      ggplot(aes(x = Disease_Group,
             y = .resid,
             fill = Disease_Group)) +
      geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
      scale_fill_manual(values = named_palette,
                        labels = plot_labels) +
      scale_x_discrete(labels = plot_labels) +
      theme_aw +
      labs(fill = "Disease Group", x = NULL, y = "Residual editing rate")
    


 parallel_plot <- create_parallel_plot(mean_residuals_df =  lm_results_sites$mean_residuals_df, 
                                       disease = disease, 
                                       feature = "locus", 
                                       feature_label = "editing site") +
     labs(title = NULL)

```







```{r,  units = "cm", fig.width = 20, fig.height = 8}
## Compare baseline to change  for plotting
    compare_baseline_df <- compare_proportions_baseline_altered(lm_results_list = lm_results_sites,
                                                                loci_consequences = loci_consequences,
                                                                factor_list = contrast_list)
    genic_dif_plot <- plot_compare_proportions_baseline(compare_proportions_df = compare_baseline_df,
                                                        metric_to_plot = "gene_type",
                                                        fill_label = "Genic Location") +
        labs(subtitle = NULL)
    biotype_dif_plot <- plot_compare_proportions_baseline(compare_proportions_df = compare_baseline_df,
                                                          metric_to_plot = "biotype_plots",
                                                          fill_label = "Biotype") +
        labs(subtitle = NULL)
    
```


```{r}
compare_baseline_df %>% 
    dplyr::select(Direction, gene_type) %>% 
     dplyr::mutate(gene_type = fct_recode(gene_type, exon = "exon", other = "intron", other = "intergenic")) %>%
    droplevels() %>%
    table() %>%
    chisq.test()
```



# EWCE analysis


## Explore the top edited genes

There are 12,017 total genes in the dataset

```{r}
common_loci <- find_sites_loci_common_to_x_samples(sites_df = comparison_df,
                                                       sample_counts_df = sample_counts,
                                                       no_common_samples_per_site = number_common_samples_per_group )
    
    sites_subset <- filter_sites_by_locus_vector(sites_df = comparison_df,
                                                 vector_of_loci = common_loci,
                                                 factor_list = contrast_list)
    
    
    number_sites_per_sample <- sites_subset %>% 
      count(Sample) %>% 
      dplyr::rename(n_sites_per_sample = n)
    
    sites_subset <- sites_subset %>% 
      left_join(number_sites_per_sample,
                by = "Sample")
    
    
    ###  Regression by sites #### 
    lm_results_sites <- regress_editing(sites_df = sites_subset,
                                        editing_metric = "Edits",
                                        covariates = covariates_used,
                                        factor_list = contrast_list,
                                        outcome_measure = outcome_to_explore,
                                        disease = disease)


dif_edits <- lm_results_sites$mean_residuals_df %>% 
  left_join(loci_consequences,
            by = "locus") %>% 
  arrange(desc(edit_difference_per_feature)) 


top_edits_sites_list <- dif_edits %>% 
  distinct(gene_name) %>% 
    drop_na() %>% 
  unlist()

top_edits_sites_id_list <- dif_edits %>% 
  distinct(gene_id) %>% 
      drop_na() %>% 
  unlist()

write.csv(top_edits_sites_list, file = str_c(args$output_folder,
                                             "/brain/",
                                             disease,
                                             "_top_edit_sites_list.csv"))

```



```{r}
dif_edits_genes <- dif_edits %>% 
  group_by(gene_name, gene_id, biotype_plots) %>% 
  summarise(mean_dif_edits_per_gene = mean(edit_difference_per_feature)) %>% 
  arrange(desc(mean_dif_edits_per_gene))

top_edits_genes_list <- dif_edits_genes %>% 
  ungroup() %>% 
  dplyr::select(gene_name) %>% 
  unlist()

top_edits_genes_id_list <- dif_edits_genes %>% 
  ungroup() %>% 
  dplyr::select(gene_id) %>% 
  unlist()

write.csv(top_edits_genes_list, file = str_c(args$output_folder,
                                             "/brain/",
                                             disease,
                                             "_top_edit_genes_list.csv"))
```

### EWCE analysis

```{r}
# Load internal specificity matrices
ctd_files <- 
  list.files(
    file.path(
      args$output_folder,
      "brain",
      "acta_snRNA_specificity_matrix_march2020"
    ), 
    pattern = "ctd", 
    full.names = T)
    

ctd_list <- vector(mode = "list", length = length(ctd_files))

for(i in 1:length(ctd_files)){
  
  # Load ctd
  ctd_list[[i]] <- readRDS(ctd_files[i])
  
  # Extract file name
  title <- 
    ctd_files[i] %>% 
    str_replace(".*/", "") %>% 
    str_replace("\\..*", "") %>% 
    str_replace(".*_", "")
  
  # Name list
  names(ctd_list)[i] <- title
 
}

rep_number = 10000
annot_level = 1
set.seed(234)

top_gene_n  = 50
ctd_dataset = ctd_list[[disease]]
```


#### Relative to PD Brain

Genes that contain the greatest difference in mean editing rate per gene between control and PD
```{r}
ctd_dataset = ctd_list[[disease]]
ewce_genes <- EWCE::bootstrap_enrichment_test(sct_data = ctd_dataset,
                                                sctSpecies = "human",
                                                genelistSpecies = "human", no_cores = 16,
                                                 hits = top_edits_genes_list[1:top_gene_n], bg = top_edits_genes_list,
                                                reps = rep_number,
                                                annotLevel = annot_level)
plot_list <- EWCE::ewce_plot(total_res = ewce_genes$results,
                           mtc_method = "BH",
                           ctd = ctd_dataset)
plot_list[[1]]

ewce_genes$results

```

```{r, fig.width = 5, fig.height=8}
ewce_figure <- ewce_genes$results %>% 
    dplyr::mutate(sd_from_mean = ifelse(sd_from_mean < 0, NA, sd_from_mean),
                  annotLevel = factor(annotLevel),
                  log10p = -log10(q)) %>% 
    ggplot(aes(y= CellType, x = annotLevel, size = sd_from_mean, colour = log10p)) +
    geom_point() +
    scale_x_discrete(labels = element_blank()) +
  #  scale_colour_continuous(trans = "log10") +
    scale_colour_viridis_c() +
    theme_aw +
    labs(colour = "-log10(p value)", size = "SD from mean", x = NULL)
```


#### Relative to Control Brain

Genes that contain the greatest difference in mean editing rate per gene between control and PD
```{r}
ctd_dataset = ctd_list$Control
ewce_genes_control <- EWCE::bootstrap_enrichment_test(sct_data = ctd_dataset,
                                                sctSpecies = "human",
                                                genelistSpecies = "human", no_cores = 16,
                                                  hits = top_edits_genes_list[1:top_gene_n], bg = top_edits_genes_list,
                                                reps = rep_number,
                                                annotLevel = annot_level)
plot_list <- EWCE::ewce_plot(total_res = ewce_genes_control$results,
                           mtc_method = "BH",
                           ctd = ctd_dataset)
plot_list[[1]]

ewce_genes_control$results
```


```{r, fig.width = 18, fig.height = 12}

supp_7_figure <- ((total_sites_violin +  edit_resid_violin +  parallel_plot) /
    (genic_dif_plot +  biotype_dif_plot + ewce_figure + plot_layout(widths = c(2, 2, 1.5)))) +plot_annotation(tag_levels = 'a')

supp_7_figure

ggsave(supp_7_figure,
       file = str_c(args$output_folder, "/supp_7_figure.pdf"),
       device = "pdf",
       dpi = 300,
       units = "mm",
       width = 300,
       height = 180,
       limitsize = TRUE
)

ggsave(supp_7_figure,
       file = str_c(args$output_folder, "/supp_7_figure.png"),
       path = args$figure_output_dir,
       device = "png",
       dpi = 300,
       units = "mm",
       width = 300,
       height = 180,
       limitsize = TRUE
)

                    
```

