---
title: "Processing acta-neuropath samples"
author: 
- name: "Aaron Wagen"
  affiliation: UCL
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  bookdown::html_document2:
    figure_caption: yes
    code_folding: show
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    number_sections: true
---


```{r setup, include = FALSE}

library(here) # For project-specific paths
library(tidyverse)
library(GenomicRanges)
library(plyranges)
library(paletteer)
library(patchwork)
library(paletteer)
library(ggtranscript)
library(GGally)
library(clusterProfiler)
library(imager)
library(broom)
library(ggforce)
library(tidycat)
library(performance)
library(MASS)


knitr::opts_chunk$set(echo = T, warning = F, message = F)

# Set defaults for ggplots 
theme_aw <- theme_bw(base_family = "Helvetica",
           base_size = 12) + 
  theme(panel.grid.major.x = element_blank(),
        legend.position = "right",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.y = element_text(vjust = 0.6),
        panel.spacing = unit(0.1, "lines"))


# Load functions
source(here::here("r_scripts",
                  "r_functions.R"))

source(here::here("r_scripts",
                  "final_editing_functions_editing_summary_metric.R"))


CORES=1

palette = c("#2a75b2", "grey75", "#B0C6E7", "deeppink3")
#palette = c("cyan4", "grey75")

```

```{r}
args <- list(
  # alu_annotations_range_file = here::here("raw_data",
  #                                         "acta_neuropath_bulk",
  #                                         "editing",
  #                                         "vep",
  #                                         "alu_annotations_distinct_range_230126.rds"),
  # reduced_ensemble_range_file = here::here("raw_data",
  #                                          "acta_neuropath_bulk",
  #                                          "editing",
  #                                          "vep",
  #                                          "reduced_ensembl_38-93_range_230126_stranded_distinct.rds"),
  # vep_results_range_file = here::here("raw_data",
  #                                     "acta_neuropath_bulk",
  #                                     "editing",
  #                                     "vep",
  #                                     "vep_results_unique_annotations_range.RDS"),
  detect_summary_comparison_sig_file = here::here("processed_data",
                                                  "acta_neuropath_bulk",
                                                  "jacusa_detect",
                                                  "detect_summary_comparison_sig_edits.RDS"),
  detect_summary_comparison_all_file = here::here("processed_data",
                                                  "acta_neuropath_bulk",
                                                  "jacusa_detect",
                                                  "detect_summary_comparison_all_edits.RDS"),
  detect_summary_comparison_alu_file = here::here("processed_data",
                                                  "acta_neuropath_bulk",     
                                                  "jacusa_detect",
                                                  "detect_summary_comparison_sig_alu_edits.RDS"),
  sample_annotation_file = here::here("raw_data",
                                      "acta_neuropath_bulk",
                                      "SampleAnnot.txt"),
  dif_summary_comparison_sig_file = here::here("processed_data",
                                               "acta_neuropath_bulk",
                                               "differential_summary_comparison_sig_edits.RDS"),
  dif_summary_comparison_all_file = here::here("processed_data",
                                               "acta_neuropath_bulk",
                                               "differential_summary_comparison_all_edits.RDS"),
  qc_results_dir = here::here("processed_data",
                              "acta_neuropath_bulk",
                              "astro_neuro_editing_QC"),
  RNA_location_file = here::here("references",
                                 "All_RNA_subcellular_localization_data.txt"),
  panel_app_file = here::here("references",
                              "panel_app_PD_complex_park_gene_list.tsv"),
  output_folder = here::here("processed_data",
                             "acta_neuropath_bulk"),
  gencode26_lengths_summary_path = here::here("references",
                                              "gencode26_gene_lengths_summary.RDS")
)




plot_labels = c("Decreased_editing" = "Decreased editing",
                "Increased_editing" = "Increased editing",
                "Lost_edit_sites" = "Lost editing sites",
                "New_edit_sites" = "New editing sites",
                "astro_nt_vs_asyn" = "Astrocyte untreated vs a-syn",
                "neuron_nt_vs_asyn" = "Neuron untreated vs a-syn",
                "cocult_nt_vs_asyn" = "Coculture untreated vs a-syn",
                "astro_nt" = "Astrocyte untreated",
                "astro_asyn" = "Astrocyte a-syn",
                "cocult_nt" = "Coculture untreated",
                "cocult_asyn" = "Coculture a-syn",
                "neuron_nt" = "Neuron untreated",
                "neuron_asyn" = "Neuron a-syn",
                "exon" = "Exon",
                "intron" = "Intron",
                "intergenic" = "Intergenic",
                "not_significant" = "Not significant",
                "significant" = "Significant",
                "Astrocyte_increase" = "Astrocyte increase",
                "Astrocyte_decrease" = "Astrocyte decrease",
                "Neuron_increase" = "Neuron increase",
                "Neuron_decrease" = "Neuron decrease",
                "Coculture_increase" = "Coculture increase",
                "Coculture_decrease" = "Coculture decrease",
                "no_treatment" = "Baseline",
                "increase" = "Increase",
                "decrease" = "Decrease",
                "increased_sites" = "Increased editing rate",
                "decreased_sites" = "Decreased editing rate",
                "treated_edits" = "a-syn",
                "untreated_edits" = "Untreated"
                )

plot_labeller = as_labeller(plot_labels)



named_palette = c(
  increased_sites = "deeppink3",
  increased = "deeppink3",
  new_sites = "hotpink1",
  decreased_sites = "#2a75b2",
  decreased = "#2a75b2",
  lost_sites = "lightskyblue",
  astro_nt = "#7FDEEA",
  astro_asyn = "#0097A6",
  neuron_nt = "#FF91C8",
  neuron_asyn = "deeppink3",
  cocult_nt = "#9FA7D8",
  cocult_asyn = "#303F9F",
  Control = "#1d9064",
  no_treatment = "#1d9064",
  DLB = "#ce490a",
  PD  = "#625aa4",
  `1uM_syn_oligomer`= "#625aa4",
  PDD = "#de0077"
)

```


```{r}
disease_groups = "PD"
disease = "PD"
factor_levels = c("Control", "PD") # Always list factor levels with control/untreated first.
covariate_list = c("RIN", "Sex")
outcome_to_explore = "Disease_Group"
factor_list <- list(Control = "Control",   # The control factor should be first
                    PD = "PD")
#disease_groups = c("DLB", "PD", "PDD")

```


```{r}
#disease_groups = c("DLB", "PD", "PDD")
disease_groups = "1uM_syn_oligomer"
disease = "1uM_syn_oligomer"
covariate_list = vector() # No covariates used here
factor_levels = c("no_treatment", "1uM_syn_oligomer") # Always list factor levels with control/untreated first.
outcome_measure = "Disease_Group" # In this case easier to relable the treatment column
outcome_to_explore = "Disease_Group"
factor_list <- list(no_treatment = "no_treatment",   # The control factor should be first
                    `1uM_syn_oligomer` = "1uM_syn_oligomer")
```



> Aim: This notebook explores variations of a gene-based summary metric of editing. Instead of a site-based metric, such an approach would allow integraton of gene based editing information with the plethora of other gene based metrics available.
<br><br>



# Methods 

## Identifying editing sites

RNA editing analysis was undertaken with JACUSA2. This utilises a drichilet multinomial distribution to ascertain whether transcripts are edited at a genomic site in two modes: in ‘detect’ mode it compares transcripts against a reference genome identifying editing in individual samples; in ‘differential’ mode it will compare two samples against each other, looking at sites that are differentially edited in one sample compared to another. Noting that transcripts with increased editing might not be successfully mapped during the alignment step, multi-sample 2 pass mapping with STAR was rerun, allowing a more generous mismatch rate of 16 base pairs per 100. This did not increase the rate of multimapping during alignment.
<br>
Each sample was input to JACUSA2 detect, to output a list of editing sites for each of 24 samples. Significant editing sites were considered with a z-score > |1.96|. 


## Summarising editing at site level verse gene level


**Editing per site**

JACUSA outputs 2 metrics:

1. The position of editing sites.
2. The editing index/rate: the number of reads that are edited divided by the total number of reads at a site.

These results were analysed directly on a per site basis. They were annotated with a GTF annotation file, alu regions from repeat masker, using ensembl VEP to explore the downstream consequences of editing sites, filtered for the most severe consequence.


**Editing per gene**

From these results per gene metrics can be derived. It will have 3 components:

1. A measure of editing rate per gene:
  1.1 Mean editing rate per gene: the mean of all the editing rates at sites in a gene
  1.2 Median editing rate per gene
  1.3 Mode of editing rate per gene
  1.4 Minimum editing rate per gene
  1.5 Maximum editing rate per gene
2. Number of editing sites in a gene
3. A variable to adjust for gene length - to account for the increased likelihood of editing with an increased number of adenosines, alus etc
  3.1 Total exon length in a gene
  3.2 Total CDS length
  3.3 Total 3'UTR length
  3.4 Alu density

**Deriving an editing summary metric**
$$
\frac{metric \ of \ editing \ rate \ per \ gene \times number \ of \ editing \ sites}{metric \ of \ gene \ length}
$$

<br>

**Adjusted editing summary metric**: The result of the above metric adjusted by highest overall gene metric value so all values are between 0 and 1.


**Other metrics to explore**

* Editing density - this would be a measure of proportion of editing sites in a region.
  * In the first instance this can be editing density per gene (adjusted by gene length). 
  * Beyond that, to look at specific regions of hyperediting, could use a sliding window approach. The question is what window size/density threshold to set (or what output measure would best test which threshold to use). Note that editing per gene does not capture this as gene length varies so much. One challenge of this approach is that it may just select for alu regions, and that the data here is polyA selected.


## Covariate correction using linear regression

In order to correct for covariates, metrics from individuals were input into a linear regression model along with covariates.


## Covariate correction using linear regression

In order to correct for covariates, metrics from individuals were input into a linear regression model along with covariates. The model is:

$$
{Editing \ metric \sim \ Disease/Treatment \ Group + \ covariates}
$$


**Editing per gene (Adjusted covariate-corrected editing summary metric)**: The residuals of the linear regression can be taken to give a covariate corrected editing metric per gene. These can optionally be converted back to their original scale.

- Question: whether to use .resid (studentised residuals) or .std.resid (standardised residuals). Whether to manipulate these at all after derivation, or use them as is?

**Gene-based editing per sample**: The mean of the residuals for each sample are then combined to create a mean, covariate corrected, gene-based editing metric per sample.

**Difference in editing per gene**: Difference in case/treatment verse control. This can be adjusted to a z score.



## Hypotheses for analysis of brain data:

| Metric to use               | Analyses                                           |
|-----------------------------|---------------------------------------------------------|
| Using per site data:        | - Are there any broad changes of change in editing site number of editing rate between case and control. |
|                             | - Do any editing sites with a significant/severe consequence change between case and control?          |
|                             | - Do editing sites match specific genomic regions? Eg: 3'UTR, Alus, ARE rich regions?  |
|                             | - Do edits correspond with mutations that increase/decrease risk of PD (whether monogenic mutations or SNPs)? |
| Using gene level data:      | - Descriptive: Which genes are most highly edited?    |
|                             | - Which genes are outliers in the distribution of edited genes: high or low editing |
|                             | - Which genes show the greatest difference between case and control?            |
|                             | - Correlations between editing and gene expression?    |
| Using editing density data:   | - In which genes/regions to hyperedited regions lie? |
|                               | - What is the correlation of genes with hyperediting in terms of expression/function?  |




## Regarding p values

* It would be possible to generate a p value for the gene based metric by converting the Z score of editing at individual sites to p values, and then combining these p-values using an existing method, eg the [harmonic mean p value](https://en.wikipedia.org/wiki/Harmonic_mean_p-value), an extension of [Fisher's method](https://en.wikipedia.org/wiki/Fisher%27s_method) but where p-values are dependent. 
* However, in using this method to generate a single p value per gene, that p value will have no relationship with the summary editing metric per gene, and the resulting distribution of these summary metrics per gene. It may be better to look a the distribution of editing metrics and look at those at the extremes (top and bottom) of the distribution. Given this it is better to use the p values to select a given site, and the summary metric to generate a distribution with which to see whether genes are significant or no.
<br>

## Additional options for consideration

* Returning residuals to original scale (in this case, that would just be the adjusted summary metric scale (0-1))
* Methods to avoid SNPs: using individual genomes were available, or filtering out known A-to-I SNP sites from editing sites.
* How consistent must an editing site be across samples for it to be considered valid:
  * One option here 
* Should the editing metric be generated using all sites in a gene (regardless of significance) or just the significant sites.


## Datasets used:

This notebook explores the previously published acta-neuropathological dataset from [Reynolds et al](https://doi.org/10.1007/s00401-021-02343-x), which used post-mortem brain data from cases with PD, PDD and DLB, and those of controls. For the purpose of this analysis, I will use only the PD verse control samples.

# Supplementary code {.tabset}

## Processing Annotation files

Note that the alu annotations are unstranded, while the gtf and vep results are stranded.

These have been processed as per the following scripts and documents.

```{r, eval = F}
#source(here::here("r_scripts", "process_vep_results.R"))

```


## Processing and annotating detect data


The processing script can be sourced directly with the following code:
```{r, eval = F}
#source(here::here("r_scripts", "process_detect_acta.R"))
```


## Processing and annotating differential data

The processing script can be sourced directly with the following code:

```{r, eval = F}
#source(here::here("r_scripts", "process_differential_data_acta.R"))
```


The detailed order for filtering is below:

```{r}
source(here::here("r_scripts",
                  "filtering_args.R"))

```



# Results


## Background data re samples

Sample annotation and information

```{r}
# Load annotation file
sample_info <- read_delim(args$sample_annotation_file)
sample_info <- sample_info %>% 
   dplyr::select(Sample = sample_name, Group = Disease_Group, Sex, AoO, AoD, DD, PMI, RIN = RINe_bulkRNA_Tapestation, aSN, TAU, AB = `thal AB`, aCG_aSN_score = `aCG aSN score`) %>% 
  dplyr::filter(Group %in% c("Control", "PD"))
```


Samples per disease group

```{r}
sample_counts <- sample_info %>% 
  count(Disease_Group = Group)
sample_counts
```

```{r, fig.width = 16, fig.height = 7}
RIN_bar <- sample_info %>% 
  ggplot(aes(x=Sample, y = RIN, fill = Group)) +
  geom_bar(stat = "identity") +
  theme_aw +
    scale_fill_manual(values = named_palette) + 
  labs(title = "Samples RIN")
  

AOD_bar <- sample_info %>% 
  ggplot(aes(x=Sample, y = AoD, fill = Group)) +
  geom_bar(stat = "identity") +
  theme_aw +
      scale_fill_manual(values = named_palette) + 
  labs(title = "Age of death")

PMI_bar <- sample_info %>% 
  ggplot(aes(x=Sample, y = PMI, fill = Group)) +
  geom_bar(stat = "identity") +
  theme_aw +
      scale_fill_manual(values = named_palette) + 
  labs(title = "Post mortem interval")

sex_bar <- sample_info %>% 
  ggplot(aes(x=Group, fill = Sex)) +
  geom_bar(stat = "count",
           position = "stack") +
  theme_aw +
      scale_fill_paletteer_d("werpals::pan",
                        direction = -1) +
  labs(title = "Samples: sex")



(RIN_bar | AOD_bar | PMI_bar | sex_bar | plot_layout(guides = "collect", widths = c(2,2,2,1)))

```


## Editing data - approach

3 different editing site groups will be considered as follows:

* 1) Sites common to all 24 samples (6004 sites)
* 2) Sites common to at least 4 samples of each disease group, and that are common between all groups ()
* 3) All editing sites (1,083,830 sites) - not used in regression analysis
* Sites where all the samples of a disease vs control pairing were the same throughout that locus.

To explore relationships between the diseases vs control the following steps were taken:

* Linear regression with Editing rate as the outcome variable and Disease Group as the predictor, adjusted for RIN, Age at death and post mortem interval.
  * ie: lm(Edits ~ Disease_Group + RIN + AoD  + PMI)
  * Sex was not adjusted for given the samples were not matched for sex
  * A negative binomial regression could not be undertaken given that the Editing metric is already adjusted for total number of reads at the site, and therefore not a count metric, but a fraction. To use a negative binomial (as per a formal differential gene expression analysis) would need to adjust the Editing metric to take into account read depth at a site, but not as a fraction.
* The residuals of the output were graphed, and then, given they were non-linear, compared statistically with a Kruskal Wallis test



```{r}
# Load file of significant results
comparison_df <- readRDS(args$detect_summary_comparison_sig_file)

# Replace NAs in genic type with intergenic, and create column of collapsed biotypes for plotting
comparison_df <- comparison_df %>% 
  tidyr::replace_na(list(gene_type = "intergenic")) %>% 
  dplyr::mutate(biotype_plots = gene_biotype) %>% 
  dplyr::filter(Disease_Group %in% c("PD", "Control")) 


comparison_df$Disease_Group <- factor(comparison_df$Disease_Group)

levels(comparison_df$biotype_plots) <- vep_levels_BIOTYPE
comparison_df$locus <- factor(comparison_df$locus)

# Check levels in data match levels in factor to be 
dataframe_levels <- levels(comparison_df$Disease_Group)
supplied_levels <- unlist(factor_list)
if (!all(dataframe_levels %in% supplied_levels) && all(supplied_levels %in% dataframe_levels)) {
  errorCondition("Supplied sorting factor does not match levels in data")
}
levels(comparison_df$Disease_Group) <- factor_list


comparison_df %>% arrange(locus)

```

## Editing metric summary per site and gene


12,492 sites out of 836,237 total sites are present in all 12 samples. As per the historgram below there is an exponential decline in the number of sites appearing in 

```{r}
sites_across_samples <- comparison_df %>% 
  group_by(locus) %>% 
  summarise(num_samples = n_distinct(Sample)) %>% # Count the number of distinct samples per locus
  arrange(desc(num_samples))


sites_across_samples %>% 
  ggplot(aes(x = num_samples)) +
  geom_histogram(bins = 12) +
  scale_x_continuous(breaks = 1:12, ) +
  theme_aw + 
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5)) +
  labs(title = "Instances of editing sites across the 24 samples",
       x = "Instances of an editing site")

```

## Comparing editing rate per site to gene summary metric



```{r}

gene_lengths <- readRDS(args$gencode26_lengths_summary_path)
gene_lengths <- remove_suffix_from_gene_id(gene_lengths, column = "gene_id")

editing_metrics <- c("Edits", "edit_adjusted")


```


### All sites

```{r, fig.width = 28, fig.height = 6}
site_label <- "all_sites"
sites_subset <- copy(comparison_df)
site_annotation <- "All sites: "

number_loci <- length(unique(comparison_df$locus))

# edit_summary_df_2sites <- edit_summary_df %>%
#   dplyr::filter(no_sites >1)

edit_summary_df <- sites_subset %>%
   get_editing_summary_metric(summary_df = .,
                                 gene_lengths_df = gene_lengths,
                              outcome_measure = outcome_measure,
                              covariates = covariate_list) %>%
      group_by(gene_name) %>%
      dplyr::filter(!all(edit_mean_adjusted == first(edit_mean_adjusted)))

site_summary_plot <- wrap_elements(create_editing_summary_plot_series(sites_subset_df = sites_subset,
                                                                 edit_summary_df = edit_summary_df) +
                                plot_annotation(str_c(site_annotation, number_loci, " loci")))


locus_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = sites_subset,
                                                                  editing_metric = "Edits",
                                                                  covariates = covariate_list,
                                                                  factor_list = factor_list,
                                                                  outcome_measure = outcome_measure,
                                                                  disease = disease) +
                                         plot_annotation(str_c(site_annotation, "regression of editing rate per site loci")))

gene_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = edit_summary_df,
                           editing_metric = "edit_mean_adjusted",
                           covariates = covariate_list,
                           outcome_measure = outcome_measure,
                           factor_list = factor_list,
                           disease = disease)  +
                             plot_annotation(str_c(site_annotation, "regression of adjusted mean metric per gene")))


assign(str_c(site_label, "_site_summary_plot"), # Rename the summary plot
       site_summary_plot)
assign(str_c(site_label, "_locus_regression_plot"), # Rename the summary plot
       locus_regression_plot)
assign(str_c(site_label, "_gene_regression_plot"), # Rename the summary plot
       gene_regression_plot)

all_sites_site_summary_plot
all_sites_gene_regression_plot
all_sites_locus_regression_plot
```

### Sites common to 1 sample per group

```{r, fig.width = 28, fig.height = 6}

number_common_samples_per_group = 1
site_label <- "common_1_sample"
site_annotation <- "All sites common to treatment and control: "
sites_to_explore <- copy(comparison_df)

common_loci <- find_sites_loci_common_to_x_samples(sites_df = sites_to_explore,
                                    sample_counts_df = sample_counts,
                                    no_common_samples_per_site = number_common_samples_per_group )

sites_subset <- filter_sites_by_locus_vector(sites_df = sites_to_explore,
                             vector_of_loci = common_loci,
                             factors_between_groups = factor_list)

edit_summary_df <- sites_subset %>% 
   get_editing_summary_metric(summary_df = .,
                                 gene_lengths_df = gene_lengths,
                              outcome_measure = outcome_to_explore,
                              covariates = covariate_list) %>% 
      group_by(gene_name) %>% 
      dplyr::filter(!all(edit_mean_adjusted == first(edit_mean_adjusted)))

site_summary_plot <- wrap_elements(create_editing_summary_plot_series(sites_subset_df = sites_subset,
                                                                 edit_summary_df = edit_summary_df) +
                                plot_annotation(str_c(site_annotation, length(common_loci), " loci")))


locus_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = sites_subset,
                                                                  editing_metric = "Edits",
                                                                  covariates = covariate_list,
                                                                  factor_list = factor_list,
                                                                  outcome_measure = outcome_measure,
                                                                  disease = disease) +
                                         plot_annotation(str_c(site_annotation, "regression of editing rate per site loci")))

gene_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = edit_summary_df,
                           editing_metric = "edit_mean_adjusted",
                           covariates = covariate_list,
                           outcome_measure = outcome_measure,
                           factor_list = factor_list,
                           disease = disease)  +
                             plot_annotation(str_c(site_annotation, "regression of adjusted mean metric per gene")))



assign(str_c(site_label, "_site_summary_plot"), # Rename the summary plot
       site_summary_plot)
assign(str_c(site_label, "_locus_regression_plot"), # Rename the summary plot
       locus_regression_plot)
assign(str_c(site_label, "_gene_regression_plot"), # Rename the summary plot
       gene_regression_plot)


common_1_sample_site_summary_plot
common_1_sample_gene_regression_plot
common_1_sample_locus_regression_plot
```



### Sites common to 2 samples per group

```{r, fig.width = 28, fig.height = 6}

sites_to_explore <- copy(comparison_df)
number_common_samples_per_group = 2
site_label <- "common_2_sample"
site_annotation <- "Sites common to 2 samples per group: "

common_loci <- find_sites_loci_common_to_x_samples(sites_df = sites_to_explore,
                                    sample_counts_df = sample_counts,
                                    no_common_samples_per_site = number_common_samples_per_group )

sites_subset <- filter_sites_by_locus_vector(sites_df = sites_to_explore,
                                             vector_of_loci = common_loci,
                                             factors_between_groups = factor_list)

edit_summary_df <- sites_subset %>% 
   get_editing_summary_metric(summary_df = .,
                                 gene_lengths_df = gene_lengths,
                              outcome_measure = outcome_to_explore,
                              covariates = covariate_list) %>% 
      group_by(gene_name) %>% 
      dplyr::filter(!all(edit_mean_adjusted == first(edit_mean_adjusted)))

site_summary_plot <- wrap_elements(create_editing_summary_plot_series(sites_subset_df = sites_subset,
                                                                 edit_summary_df = edit_summary_df) +
                                plot_annotation(str_c(site_annotation, length(common_loci), " loci")))


locus_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = sites_subset,
                                                                  editing_metric = "Edits",
                                                                  covariates = covariate_list,
                                                                  factor_list = factor_list,
                                                                  outcome_measure = outcome_measure,
                                                                  disease = disease) +
                                         plot_annotation(str_c(site_annotation, "regression of editing rate per site loci")))

gene_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = edit_summary_df,
                           editing_metric = "edit_mean_adjusted",
                           covariates = covariate_list,
                           outcome_measure = outcome_measure,
                           factor_list = factor_list,
                           disease = disease)  +
                             plot_annotation(str_c(site_annotation, "regression of adjusted mean metric per gene")))



assign(str_c(site_label, "_site_summary_plot"), # Rename the summary plot
       site_summary_plot)
assign(str_c(site_label, "_locus_regression_plot"), # Rename the summary plot
       locus_regression_plot)
assign(str_c(site_label, "_gene_regression_plot"), # Rename the summary plot
       gene_regression_plot)


common_2_sample_site_summary_plot
common_2_sample_gene_regression_plot
common_2_sample_locus_regression_plot

```


### Sites common to 3 samples per group

```{r}
gene_lengths %>% 
  dplyr::filter(gene_name == "AC002310")
```


```{r}
edit_summary_df %>% arrange(gene_name)
edits_per_gene %>%  arrange(gene_name)
sites_df = edits_per_gene


```



```{r, fig.width = 28, fig.height = 6}

sites_to_explore <- copy(comparison_df)
number_common_samples_per_group = 3
site_label <- "common_3_sample"
site_annotation <- "Sites common to 3 samples per group: "

common_loci <- find_sites_loci_common_to_x_samples(sites_df = sites_to_explore,
                                    sample_counts_df = sample_counts,
                                    no_common_samples_per_site = number_common_samples_per_group )

sites_subset <- filter_sites_by_locus_vector(sites_df = sites_to_explore,
                                             vector_of_loci = common_loci,
                                             factors_between_groups = factor_list)

edit_summary_df <- sites_subset %>% 
  get_editing_summary_metric(summary_df = .,
                             gene_lengths_df = gene_lengths,
                             outcome_measure = outcome_to_explore,
                             covariates = covariate_list) %>% 
  group_by(gene_name) %>% 
  dplyr::filter(!all(edit_mean_adjusted == first(edit_mean_adjusted)))

site_summary_plot <- wrap_elements(create_editing_summary_plot_series(sites_subset_df = sites_subset,
                                                                      edit_summary_df = edit_summary_df) +
                                     plot_annotation(str_c(site_annotation, length(common_loci), " loci")))


locus_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = sites_subset,
                                                                  editing_metric = "Edits",
                                                                  covariates = covariate_list,
                                                                  factor_list = factor_list,
                                                                  outcome_measure = outcome_measure,
                                                                  disease = disease) +
                                         plot_annotation(str_c(site_annotation, "regression of editing rate per site loci")))

gene_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = edit_summary_df,
                                                                 editing_metric = "edit_mean_adjusted",
                                                                 covariates = covariate_list,
                                                                 outcome_measure = outcome_measure,
                                                                 factor_list = factor_list,
                                                                 disease = disease)  +
                                        plot_annotation(str_c(site_annotation, "regression of adjusted mean metric per gene")))



assign(str_c(site_label, "_site_summary_plot"), # Rename the summary plot
       site_summary_plot)
assign(str_c(site_label, "_locus_regression_plot"), # Rename the summary plot
       locus_regression_plot)
assign(str_c(site_label, "_gene_regression_plot"), # Rename the summary plot
       gene_regression_plot)


common_3_sample_site_summary_plot
common_3_sample_gene_regression_plot
common_3_sample_locus_regression_plot

```

### Sites common to 5 samples per group

```{r, fig.width = 28, fig.height = 6}

sites_to_explore <- copy(comparison_df)
number_common_samples_per_group = 5
site_label <- "common_5_sample"
site_annotation <- "Sites common to 5 samples per group: "

common_loci <- find_sites_loci_common_to_x_samples(sites_df = sites_to_explore,
                                    sample_counts_df = sample_counts,
                                    no_common_samples_per_site = number_common_samples_per_group )

sites_subset <- filter_sites_by_locus_vector(sites_df = sites_to_explore,
                                             vector_of_loci = common_loci,
                                             factors_between_groups = factor_list)

edit_summary_df <- sites_subset %>% 
  get_editing_summary_metric(summary_df = .,
                             gene_lengths_df = gene_lengths,
                             outcome_measure = outcome_to_explore,
                             covariates = covariate_list) %>% 
      group_by(gene_name) %>% 
      dplyr::filter(!all(edit_mean_adjusted == first(edit_mean_adjusted)))

site_summary_plot <- wrap_elements(create_editing_summary_plot_series(sites_subset_df = sites_subset,
                                                                 edit_summary_df = edit_summary_df) +
                                plot_annotation(str_c(site_annotation, length(common_loci), " loci")))


locus_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = sites_subset,
                                                                  editing_metric = "Edits",
                                                                  covariates = covariate_list,
                                                                  factor_list = factor_list,
                                                                  outcome_measure = outcome_measure,
                                                                  disease = disease) +
                                         plot_annotation(str_c(site_annotation, "regression of editing rate per site loci")))

gene_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = edit_summary_df,
                           editing_metric = "edit_mean_adjusted",
                           covariates = covariate_list,
                           outcome_measure = outcome_measure,
                           factor_list = factor_list,
                           disease = disease)  +
                             plot_annotation(str_c(site_annotation, "regression of adjusted mean metric per gene")))



assign(str_c(site_label, "_site_summary_plot"), # Rename the summary plot
       site_summary_plot)
assign(str_c(site_label, "_locus_regression_plot"), # Rename the summary plot
       locus_regression_plot)
assign(str_c(site_label, "_gene_regression_plot"), # Rename the summary plot
       gene_regression_plot)


common_5_sample_site_summary_plot
common_5_sample_gene_regression_plot
common_5_sample_locus_regression_plot

```

### Sites common to all samples

```{r, fig.width = 30, fig.height = 6}

sites_to_explore <- copy(comparison_df)
site_label <- "common_all_samples"
site_annotation <- "Sites common to all samples: "

# Filter sites for those present in all samples
common_loci <- sites_across_samples %>% 
  filter(num_samples == 12) %>% 
  pull(locus)

sites_subset <- filter_sites_by_locus_vector(sites_df = sites_to_explore,
                                             vector_of_loci = common_loci,
                                             factors_between_groups = factor_list)

edit_summary_df <- sites_subset %>% 
   get_editing_summary_metric(summary_df = .,
                                 gene_lengths_df = gene_lengths,
                              outcome_measure = outcome_to_explore,
                              covariates = covariate_list) %>% 
      group_by(gene_name) %>% 
      dplyr::filter(!all(edit_mean_adjusted == first(edit_mean_adjusted)))

site_summary_plot <- wrap_elements(create_editing_summary_plot_series(sites_subset_df = sites_subset,
                                                                 edit_summary_df = edit_summary_df) +
                                plot_annotation(str_c(site_annotation, length(common_loci), " loci")))


locus_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = sites_subset,
                                                                  editing_metric = "Edits",
                                                                  covariates = covariate_list,
                                                                  factor_list = factor_list,
                                                                  outcome_measure = outcome_measure,
                                                                  disease = disease) +
                                         plot_annotation(str_c(site_annotation, "regression of editing rate per site loci")))

gene_regression_plot <- wrap_elements(linear_regression_and_plot(sites_df = edit_summary_df,
                           editing_metric = "edit_mean_adjusted",
                           covariates = covariate_list,
                           outcome_measure = outcome_measure,
                           factor_list = factor_list,
                           disease = disease)  +
                             plot_annotation(str_c(site_annotation, "regression of adjusted mean metric per gene")))



assign(str_c(site_label, "_site_summary_plot"), # Rename the summary plot
       site_summary_plot)
assign(str_c(site_label, "_locus_regression_plot"), # Rename the summary plot
       locus_regression_plot)
assign(str_c(site_label, "_gene_regression_plot"), # Rename the summary plot
       gene_regression_plot)


common_all_samples_site_summary_plot
common_all_samples_gene_regression_plot
common_all_samples_locus_regression_plot
```




```{r, eval = T, fig.width=28, fig.height=28}

editing_metric_site_summary_plot <- (all_sites_site_summary_plot  /
                                       common_1_sample_site_summary_plot /
                                       common_2_sample_site_summary_plot /
                                       common_3_sample_site_summary_plot  /
                                       common_5_sample_site_summary_plot  /
                                       common_all_samples_site_summary_plot)

ggsave(editing_metric_site_summary_plot,
       file = here::here("manuscript/figures/acta_editing_metric_site_summary_plot.png"),
       device = "png",
       units = "cm",
       height = 58,
       width = 58)

```


## Effect of editing metric on regression results: editing rate
```{r, eval = T, fig.width=28, fig.height=28}

editing_locus_regression_summary_plot <- (all_sites_locus_regression_plot  /
                                            common_1_sample_locus_regression_plot /
                                            common_2_sample_locus_regression_plot /
                                            common_3_sample_locus_regression_plot  /
                                            common_5_sample_locus_regression_plot  /
                                            common_all_samples_locus_regression_plot)

ggsave(editing_locus_regression_summary_plot ,
       file = here::here("manuscript/figures/acta_editing_locus_regression_summary_plot.png"),
       device = "png",
       units = "cm",
       height = 68,
       width = 58)

```


## Effect of editing metric on regression results: gene metric
```{r, eval = T, fig.width=28, fig.height=28}

editing_gene_regression_summary_plot <- (all_sites_gene_regression_plot  /
                                           common_1_sample_gene_regression_plot /
                                           common_2_sample_gene_regression_plot /
                                           common_3_sample_gene_regression_plot  /
                                           common_5_sample_gene_regression_plot  /
                                           common_all_samples_gene_regression_plot)

ggsave(editing_gene_regression_summary_plot,
       file = here::here("manuscript/figures/acta_editing_gene_regression_summary_plot.png"),
       device = "png",
       units = "cm",
       height = 68,
       width = 58)

```






# Session info

<details>
  <summary>Show/hide</summary>

```{r reproducibility, echo = FALSE}
# Session info
library("sessioninfo")
options(width = 120)
session_info()
```

</details> 
